{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os\n",
    "from importlib import reload\n",
    "\n",
    "import time\n",
    "import popDMS\n",
    "\n",
    "\n",
    "# GLOBAL VARIABLES\n",
    "\n",
    "## Pointers to data and output\n",
    "OUT_DIR  = 'output/'                                         # general directory for output\n",
    "FREQ_DIR = os.path.join(OUT_DIR, 'variant_frequencies/')     # directory for experimental variant (codon/aa) frequencies\n",
    "SEL_DIR  = os.path.join(OUT_DIR, 'selection_coefficients/')  # directory for inferred selection coefficients\n",
    "DATA_DIR = 'data/'                                           # general directory for external data\n",
    "RAW_DIR  = os.path.join(DATA_DIR, 'raw_data/')               # unprocessed data directory\n",
    "FIG_DIR  = 'figures/'                                        # directory for figures\n",
    "\n",
    "## popDMS parameters\n",
    "CORR_CUTOFF_PCT = 0.1  # maximum fraction of drop in correlation per log10(gamma) for setting optimal regularization strength"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "root = 'Spike'\n",
    "n_reps = 2\n",
    "\n",
    "for f in [os.path.join(FREQ_DIR, '_'.join([root, 'single_aa_rep_%d.csv' % i])) for i in range(1, 1+n_reps)]:\n",
    "    df = pd.read_csv(f)\n",
    "    df.to_csv(f + '.gz', index=False, compression='gzip')\n",
    "\n",
    "for f in [os.path.join(FREQ_DIR, '_'.join([root, 'double_aa_rep_%d.csv' % i])) for i in range(1, 1+n_reps)]:\n",
    "    df = pd.read_csv(f)\n",
    "    df.to_csv(f + '.gz', index=False, compression='gzip')\n",
    "\n",
    "for f in [os.path.join(SEL_DIR, '_'.join([root, 'selection_coefficients.csv']))]:\n",
    "    df = pd.read_csv(f)\n",
    "    df.to_csv(f + '.gz', index=False, compression='gzip')\n",
    "\n",
    "for f in [os.path.join(OUT_DIR + 'merged_preference/', root+'.csv')]:\n",
    "    df = pd.read_csv(f)\n",
    "    df.to_csv(f + '.gz', index=False, compression='gzip')\n",
    "\n",
    "\n",
    "root = 'H3N2 NA'\n",
    "n_reps = 2\n",
    "\n",
    "for f in [os.path.join(FREQ_DIR, '_'.join([root, 'single_aa_rep_%d.csv' % i])) for i in range(1, 1+n_reps)]:\n",
    "    df = pd.read_csv(f)\n",
    "    for df_iter, df_entry in df.iterrows():\n",
    "        if df_entry['aa']=='-':\n",
    "            df.loc[df_iter, 'aa'] = '*'\n",
    "    df.to_csv(f + '.gz', index=False, compression='gzip')\n",
    "\n",
    "for f in [os.path.join(SEL_DIR, '_'.join([root, 'selection_coefficients.csv']))]:\n",
    "    df = pd.read_csv(f)\n",
    "    df.to_csv(f + '.gz', index=False, compression='gzip')\n",
    "\n",
    "for f in [os.path.join(OUT_DIR + 'merged_preference/', root+'.csv')]:\n",
    "    df = pd.read_csv(f)\n",
    "    df.to_csv(f + '.gz', index=False, compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function compute_variant_frequencies in module popDMS:\n",
      "\n",
      "compute_variant_frequencies(name, reference_sequence_file, haplotype_counts_file, n_replicates, time_points, time_cols, freq_dir='.', with_epistasis=False, comment_char=None, format='MaveDB')\n",
      "    Computes variant (allele) frequencies, used by popDMS, from haplotype counts. By\n",
      "    default, we assume that counts are saved in MaveDB format. For more information\n",
      "    on this format, see https://www.mavedb.org/docs/mavedb/data_formats.html\n",
      "    \n",
      "    Required arguments:\n",
      "        - name: String that will be associated with saved files, serving as an\n",
      "            identifier for the data set\n",
      "        - reference_sequence_file: File path to the reference sequence, which is\n",
      "            needed to specify 'WT' variants\n",
      "        - haplotype_counts_file: Path to an existing file recording haplotype\n",
      "            frequencies\n",
      "        - n_replicates: Number of experimental replicates\n",
      "        - time_points: Time points in the data; for normalization across data sets,\n",
      "            this should be roughly the number of 'generations' between sequencing\n",
      "            samples in the experiment\n",
      "        - time_cols: A dictionary mapping between the number for each experimental\n",
      "            replicate and the column in the haplotype counts file that contains\n",
      "            count information\n",
      "    \n",
      "    Optional arguments:\n",
      "        - freq_dir (default: '.'): Path to directory where frequencies will be saved\n",
      "        - with_epistasis (default: False): If True, will compute 3- and 4-amino acid\n",
      "            frequencies for epistasis analysis\n",
      "        - comment_char (default: None): Character used as a comment in the haplotype\n",
      "            counts file; in some MaveDB files, this is '#'\n",
      "        - format (default: 'MaveDB'): Format of the input data files (options:\n",
      "            'MaveDB', 'Bloom')\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(popDMS.compute_variant_frequencies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function infer_correlated in module popDMS:\n",
      "\n",
      "infer_correlated(name, n_replicates, corr_cutoff_pct, freq_dir='.', output_dir='.', with_epistasis=False, plot_gamma=True)\n",
      "    Infer selection coefficients from data that includes correlations (counts)\n",
      "    between pairs of mutant sites. This is not possible with data that includes\n",
      "    single codon/amino acid counts only.\n",
      "    \n",
      "    Required arguments:\n",
      "        - name: String that will be associated with saved files, serving as an\n",
      "            identifier for the data set\n",
      "        - n_replicates: Number of experimental replicates\n",
      "        - corr_cutoff_pct: Cutoff on the maximum allowed drop in correlation\n",
      "            between replicates, used to determine the optimal regularization\n",
      "            strength\n",
      "    \n",
      "    Optional arguments:\n",
      "        - freq_dir (default: '.'): Directory where variant frequency data has been \n",
      "            saved; assumes data saved in the format of compute_variant_frequencies\n",
      "        - output_dir (default: '.'): Directory where selection coefficients will be \n",
      "            saved\n",
      "        - with_epistasis (default: False): Flag to infer epistatic interactions; \n",
      "            this requires correlation data for sets of three and four codons/amino \n",
      "            acids; not currently functional (implemented separately in C++ code in \n",
      "            the epistasis_inference directory)\n",
      "        - plot_gamma (default: True): Plot correlation between replicates as a \n",
      "            function of the regularization strength gamma\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(popDMS.infer_correlated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function compute_variant_frequencies_Bloom in module popDMS:\n",
      "\n",
      "compute_variant_frequencies_Bloom(name, codon_counts_files, replicates, times, freq_dir='.', error_file=None, comment_char=None)\n",
      "    Compute variant frequencies from codon counts in Bloom lab format. If an error\n",
      "    file is provided, corrects for sequencing errors.\n",
      "    \n",
      "    Required arguments:\n",
      "        - name: Name of the experiment\n",
      "        - codon_counts_files: List of paths to codon counts files\n",
      "        - replicates: List of replicate numbers\n",
      "        - times: List of time points\n",
      "        - freq_dir: Path to directory where frequencies will be saved\n",
      "    \n",
      "    Optional arguments:\n",
      "        - error_file (default: None): Path to a file recording sequence counts from\n",
      "            sequencing the reference sequence only, used to estimate sequencing\n",
      "            error rates and correct variant frequencies\n",
      "        - comment_char (default: None): Character used to indicate comments in the\n",
      "            codon counts files\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(popDMS.compute_variant_frequencies_Bloom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function infer_independent in module popDMS:\n",
      "\n",
      "infer_independent(name, n_replicates, corr_cutoff_pct, freq_dir='.', output_dir='.', plot_gamma=True)\n",
      "    Infer selection coefficients from data that consists of single amino acid\n",
      "    frequencies only.\n",
      "    \n",
      "    Required arguments:\n",
      "        - name: String that will be associated with saved files, serving as an\n",
      "            identifier for the data set\n",
      "        - n_replicates: Number of experimental replicates\n",
      "        - corr_cutoff_pct: Cutoff on the maximum allowed drop in correlation\n",
      "            between replicates, used to determine the optimal regularization\n",
      "            strength\n",
      "    \n",
      "    Optional arguments:\n",
      "        - freq_dir (default: '.'): Directory where variant frequency data has been\n",
      "            saved; assumes data saved in the format of compute_variant_frequencies\n",
      "        - output_dir (default: '.'): Directory where selection coefficients will be\n",
      "            saved\n",
      "        - plot_gamma (default: True): Plot correlation between replicates as a \n",
      "            function of the regularization strength gamma\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(popDMS.infer_independent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4991\n",
      "Only one replicate, setting gamma=1\n",
      "CPU times: user 18.6 s, sys: 640 ms, total: 19.2 s\n",
      "Wall time: 2.48 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "barcode_data = dict(\n",
    "    name = 'test',\n",
    "    replicate_files = ['sample_barcode_counts_small.csv'],\n",
    "    output_dir = '.',\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_barcode(**barcode_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# YAP1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'YAP1'\n",
    "n_replicates = 2\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    reference_sequence_file = os.path.join(RAW_DIR, 'YAP1_reference_sequence.dat'),\n",
    "    haplotype_counts_file = os.path.join(RAW_DIR, 'YAP1_aa2codon_counts.csv.gz'), \n",
    "    freq_dir = FREQ_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    time_points = [0, 1, 2, 3],\n",
    "    time_cols = {1: ['101208_c_0', '101208_c_1', '101208_c_2', '101208_c_3'],  \n",
    "                 2: ['110307_c_0', '110307_c_1', '110307_c_2', '110307_c_3']},\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    with_epistasis = False,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_correlated(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TpoR MPL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "name = 'TpoR'\n",
    "n_replicates = 6\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    reference_sequence_file = os.path.join(RAW_DIR, 'TpoR_reference_sequence.dat'),\n",
    "    haplotype_counts_file = os.path.join(RAW_DIR, 'TpoR_nucleotide_counts.csv'), \n",
    "    freq_dir = FREQ_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    time_points = [0, 1],\n",
    "    time_cols = {1: ['Replicate_A_c_0', 'Replicate_A_c_1'],\n",
    "                 2: ['Replicate_B_c_0', 'Replicate_B_c_1'],\n",
    "                 3: ['Replicate_C_c_0', 'Replicate_C_c_1'],\n",
    "                 4: ['Replicate_D_c_0', 'Replicate_D_c_1'],\n",
    "                 5: ['Replicate_E_c_0', 'Replicate_E_c_1'],\n",
    "                 6: ['Replicate_F_c_0', 'Replicate_F_c_1']},\n",
    "    comment_char = '#'\n",
    ")\n",
    "popDMS.compute_variant_frequencies(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    with_epistasis = False,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_correlated(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TpoR MPL S505N"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "name = 'TpoR_S505N'\n",
    "n_replicates = 6\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    reference_sequence_file = os.path.join(RAW_DIR, 'TpoR_reference_sequence.dat'),\n",
    "    haplotype_counts_file = os.path.join(RAW_DIR, 'TpoR_S505N_nucleotide_counts.csv'), \n",
    "    freq_dir = FREQ_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    time_points = [0, 1],\n",
    "    time_cols = {1: ['Replicate_1_c_0', 'Replicate_1_c_1'],\n",
    "                 2: ['Replicate_2_c_0', 'Replicate_2_c_1'],\n",
    "                 3: ['Replicate_3_c_0', 'Replicate_3_c_1'],\n",
    "                 4: ['Replicate_4_c_0', 'Replicate_4_c_1'],\n",
    "                 5: ['Replicate_5_c_0', 'Replicate_5_c_1'],\n",
    "                 6: ['Replicate_6_c_0', 'Replicate_6_c_1']},\n",
    "    comment_char = '#'\n",
    ")\n",
    "popDMS.compute_variant_frequencies(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    with_epistasis = False,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_correlated(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ube4b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "name = 'Ube4b'\n",
    "n_replicates = 2\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    reference_sequence_file = os.path.join(RAW_DIR, 'Ube4b_reference_sequence.dat'),\n",
    "    haplotype_counts_file = os.path.join(RAW_DIR, 'Ube4b_nucleotide_counts.csv'), \n",
    "    freq_dir = FREQ_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    time_points = [0, 1, 2, 3],\n",
    "    time_cols = {1: ['Rep_2_c_0', 'Rep_2_c_1', 'Rep_2_c_2', 'Rep_2_c_3'],\n",
    "                 2: ['Rep_3_c_0', 'Rep_3_c_1', 'Rep_3_c_2', 'Rep_3_c_3']},\n",
    "    comment_char = '#'\n",
    ")\n",
    "popDMS.compute_variant_frequencies(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    with_epistasis = False,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_correlated(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BRCA1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'Y2H_1'\n",
    "n_replicates = 3\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    reference_sequence_file = os.path.join(RAW_DIR, 'BRCA1_reference_sequence.dat'),\n",
    "    haplotype_counts_file = os.path.join(RAW_DIR, 'BRCA1_nucleotide_counts.csv'), \n",
    "    freq_dir = FREQ_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    time_points = [0, 18, 37, 45],\n",
    "    time_cols = {1: ['Y2H_1_Rep1_c_0', 'Y2H_1_Rep1_c_18', 'Y2H_1_Rep1_c_37', 'Y2H_1_Rep1_c_45'],\n",
    "                 2: ['Y2H_1_Rep2_c_0', 'Y2H_1_Rep2_c_18', 'Y2H_1_Rep2_c_37', 'Y2H_1_Rep2_c_45'],\n",
    "                 3: ['Y2H_1_Rep3_c_0', 'Y2H_1_Rep3_c_18', 'Y2H_1_Rep3_c_37', 'Y2H_1_Rep3_c_45']},\n",
    "    comment_char = '#'\n",
    ")\n",
    "popDMS.compute_variant_frequencies(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    with_epistasis = False,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_correlated(**selection_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'Y2H_2'\n",
    "n_replicates = 3\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    reference_sequence_file = os.path.join(RAW_DIR, 'BRCA1_reference_sequence.dat'),\n",
    "    haplotype_counts_file = os.path.join(RAW_DIR, 'BRCA1_nucleotide_counts.csv'), \n",
    "    freq_dir = FREQ_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    time_points = [0, 16, 41, 64],\n",
    "    time_cols = {1: ['Y2H_2_Rep1_c_0', 'Y2H_2_Rep1_c_16', 'Y2H_2_Rep1_c_41', 'Y2H_2_Rep1_c_64'],\n",
    "                 2: ['Y2H_2_Rep2_c_0', 'Y2H_2_Rep2_c_16', 'Y2H_2_Rep2_c_41', 'Y2H_2_Rep2_c_64'],\n",
    "                 3: ['Y2H_2_Rep3_c_0', 'Y2H_2_Rep3_c_16', 'Y2H_2_Rep3_c_41', 'Y2H_2_Rep3_c_64']},\n",
    "    comment_char = '#'\n",
    ")\n",
    "popDMS.compute_variant_frequencies(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    with_epistasis = False,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_correlated(**selection_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'E3'\n",
    "n_replicates = 6\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    reference_sequence_file = os.path.join(RAW_DIR, 'BRCA1_reference_sequence.dat'),\n",
    "    haplotype_counts_file = os.path.join(RAW_DIR, 'BRCA1_nucleotide_counts.csv'), \n",
    "    freq_dir = FREQ_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    time_points = [0, 1, 2, 3, 4, 5],\n",
    "    time_cols = {1: ['PlusE2NewRep3_c_0', 'PlusE2NewRep3_c_1', 'PlusE2NewRep3_c_2', 'PlusE2NewRep3_c_3', 'PlusE2NewRep3_c_4', \n",
    "                     'PlusE2NewRep3_c_5'],\n",
    "                 2: ['PlusE2NewRep4_c_0', 'PlusE2NewRep4_c_1', 'PlusE2NewRep4_c_2', 'PlusE2NewRep4_c_3', 'PlusE2NewRep4_c_4', \n",
    "                     'PlusE2NewRep4_c_5'],\n",
    "                 3: ['PlusE2NewRep5_c_0', 'PlusE2NewRep5_c_1', 'PlusE2NewRep5_c_2', 'PlusE2NewRep5_c_3', 'PlusE2NewRep5_c_4', \n",
    "                     'PlusE2NewRep5_c_5'],\n",
    "                 4: ['PlusE2Rep3_c_0', 'PlusE2Rep3_c_1', 'PlusE2Rep3_c_2', 'PlusE2Rep3_c_3', 'PlusE2Rep3_c_4', 'PlusE2Rep3_c_5'],\n",
    "                 5: ['PlusE2Rep4_c_0', 'PlusE2Rep4_c_1', 'PlusE2Rep4_c_2', 'PlusE2Rep4_c_3', 'PlusE2Rep4_c_4', 'PlusE2Rep4_c_5'],\n",
    "                 6: ['PlusE2Rep5_c_0', 'PlusE2Rep5_c_1', 'PlusE2Rep5_c_2', 'PlusE2Rep5_c_3', 'PlusE2Rep5_c_4', 'PlusE2Rep5_c_5']},\n",
    "    comment_char = '#'\n",
    ")\n",
    "popDMS.compute_variant_frequencies(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    with_epistasis = False,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_correlated(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DBR1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'DBR1'\n",
    "n_replicates = 2\n",
    "\n",
    "count_data = dict(\n",
    "    dir = FREQ_DIR,\n",
    "    name = name,\n",
    "    types = ['single', 'double'],\n",
    "    n_replicates = n_replicates,\n",
    "    reference_sequence_file = os.path.join(RAW_DIR, name + '_reference_sequence.dat'),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.convert_codon_counts_to_variant_frequencies(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    with_epistasis = False,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_correlated(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SARS-CoV-2 Spike"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'Spike'\n",
    "n_replicates = 2\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_correlated(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Independent (no mutant-mutant correlations) proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Zika virus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'ZIKV'\n",
    "n_replicates = 3\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % name),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Influenza Perth 2009 (H3N2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'Perth2009'\n",
    "n_replicates = 4\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % name),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Influenza WSN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'WSN'\n",
    "n_replicates = 3\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % name),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Influenza PR8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'PR8'\n",
    "n_replicates = 3\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % name),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Influenza Aichi 68C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'Aichi68C'\n",
    "n_replicates = 2\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % name),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Influenza Matrix_M1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'Matrix_M1'\n",
    "filename = 'HomM1'\n",
    "n_replicates = 3\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % filename),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Influenza A549"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'A549'\n",
    "n_replicates = 3\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % name),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Influenza CCL141"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'CCL141'\n",
    "n_replicates = 3\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % name),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Influenza MS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'MS'\n",
    "n_replicates = 2\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % name),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Influenza MxA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'MxA'\n",
    "n_replicates = 2\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % name),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Influenza MxAneg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'MxAneg'\n",
    "n_replicates = 2\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (name, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % name),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Influenza H3N2 NA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'H3N2 NA'\n",
    "n_replicates = 2\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HIV Env BG505"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'HIV Env BG505'\n",
    "filename = 'BG505'\n",
    "n_replicates = 3\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % filename),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HIV Env BF520"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'HIV Env BF520'\n",
    "filename = 'BF520'\n",
    "n_replicates = 3\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % filename),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HIV Env BF520 human host"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'HIV BF520 human'\n",
    "filename = 'human'\n",
    "n_replicates = 2\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % filename),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HIV Env BF520 rhesus host"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'HIV BF520 rhesus'\n",
    "filename = 'rhesus'\n",
    "n_replicates = 2\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % filename),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HIV bnAbs FP16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'HIV bnAbs FP16'\n",
    "filename = 'FP16'\n",
    "n_replicates = 2\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % filename),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HIV bnAbs FP20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'HIV bnAbs FP20'\n",
    "filename = 'FP20'\n",
    "n_replicates = 2\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % filename),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HIV bnAbs VRC34"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "reload(popDMS)\n",
    "\n",
    "name = 'HIV bnAbs VRC34'\n",
    "filename = 'VRC34'\n",
    "n_replicates = 2\n",
    "\n",
    "count_data = dict(\n",
    "    name = name,\n",
    "    codon_counts_files = [os.path.join(RAW_DIR, '%s_mutDNA-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)] + \\\n",
    "                         [os.path.join(RAW_DIR, '%s_mutvirus-%d_codoncounts.csv' % (filename, i+1)) for i in range(n_replicates)],\n",
    "    replicates = [i+1 for i in range(n_replicates)] + [i+1 for i in range(n_replicates)],\n",
    "    times = [0 for i in range(n_replicates)] + [1 for i in range(n_replicates)],\n",
    "    freq_dir = FREQ_DIR,\n",
    "    # error_file = os.path.join(RAW_DIR, '%s_DNA_codoncounts.csv' % filename),\n",
    "    comment_char = None\n",
    ")\n",
    "popDMS.compute_variant_frequencies_Bloom(**count_data)\n",
    "\n",
    "selection_data = dict(\n",
    "    name = name,\n",
    "    freq_dir = FREQ_DIR,\n",
    "    output_dir = SEL_DIR,\n",
    "    n_replicates = n_replicates,\n",
    "    corr_cutoff_pct = CORR_CUTOFF_PCT,\n",
    "    plot_gamma = True\n",
    ")\n",
    "popDMS.infer_independent(**selection_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparing enrichment ratios and selection coefficients at example sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "name = 'HIV Env BG505'\n",
    "filename = 'BG505'\n",
    "n_replicates = 3\n",
    "gamma = 2e-3\n",
    "\n",
    "ex_sites = [        592,        596,         508,         525]\n",
    "ex_aas   = [['I', 'L',], ['S', 'F'], ['G', 'R', 'Y', 'H', 'F', 'K'], ['S', 'N',]]\n",
    "\n",
    "df = [pd.read_csv(os.path.join(FREQ_DIR, name + ('_single_aa_rep_%d.csv.gz' % (rep+1)))) for rep in range(n_replicates)]\n",
    "\n",
    "for i_site in range(len(ex_sites)):\n",
    "    print('%s site %d' % (name, ex_sites[i_site]))\n",
    "    print('residue\\treplicate\\tx0\\t\\tx1\\t\\tratio\\t\\td log\\t\\td logit\\t\\traw s\\t\\treg s')\n",
    "\n",
    "    for a in ex_aas[i_site]:\n",
    "        for i_rep in range(n_replicates):\n",
    "            df_temp = df[i_rep][(df[i_rep]['site']==ex_sites[i_site]) & (df[i_rep]['aa']==a)]\n",
    "            x0 = df_temp[df_temp['generation']==0].iloc[0]['frequency']\n",
    "            x1 = df_temp[df_temp['generation']==1].iloc[0]['frequency']\n",
    "            \n",
    "            xvar   = ((x0+x1)/2) * (1 - ((x0+x1)/2)) \n",
    "            ratio  = x1/x0\n",
    "            dlog   = np.log(x1/x0)\n",
    "            dlogit = np.log(x1/(1-x1))-np.log(x0/(1-x0))\n",
    "            s_raw  = (x1 - x0) / xvar\n",
    "            s_reg  = (x1 - x0) / (xvar + gamma)\n",
    "\n",
    "            res_str = a if i_rep==0 else ''\n",
    "            print('%s\\t%d\\t\\t%.3e\\t%.3e\\t%.3e\\t%.3e\\t%.3e\\t%.3e\\t%.3e' % (res_str, i_rep+1, x0, x1, ratio, dlog, dlogit, s_raw, s_reg))\n",
    "\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "name = 'Ube4b'\n",
    "n_replicates = 2\n",
    "gamma = 2.8e-04\n",
    "\n",
    "ex_sites = [             34]\n",
    "ex_aas   = [['A', 'D', 'N']]\n",
    "\n",
    "df = [pd.read_csv(os.path.join(FREQ_DIR, name + ('_single_aa_rep_%d.csv.gz' % (rep+1)))) for rep in range(n_replicates)]\n",
    "\n",
    "for i_site in range(len(ex_sites)):\n",
    "    print('%s site %d' % (name, ex_sites[i_site]))\n",
    "    print('residue\\treplicate\\tx0\\t\\tx1\\t\\tratio\\t\\td log\\t\\td logit\\t\\traw s\\t\\treg s')\n",
    "\n",
    "    for a in ex_aas[i_site]:\n",
    "        for i_rep in range(n_replicates):\n",
    "            df_temp = df[i_rep][(df[i_rep]['site']==ex_sites[i_site]) & (df[i_rep]['aa']==a)]\n",
    "            x = [df_temp[df_temp['generation']==i].iloc[0]['frequency'] for i in range(4)]\n",
    "\n",
    "            xvar   = [((x[i]+x[i+1])/2) * (1 - ((x[i]+x[i+1])/2)) for i in range(len(x)-1)] \n",
    "            ratio  = x[-1]/x[0]\n",
    "            dlog   = np.log(x[-1]/x[0])\n",
    "            dlogit = np.log(x[-1]/(1-x[-1]))-np.log(x[0]/(1-x[0]))\n",
    "            s_raw  = (x[-1] - x[0]) / np.sum(xvar)\n",
    "            s_reg  = (x[-1] - x[0]) / (np.sum(xvar) + gamma)\n",
    "\n",
    "            res_str = a if i_rep==0 else ''\n",
    "            print('%s\\t%d\\t\\t%.3e\\t%.3e\\t%.3e\\t%.3e\\t%.3e\\t%.3e\\t%.3e' % (res_str, i_rep+1, x[0], x[-1], ratio, dlog, dlogit, s_raw, s_reg))\n",
    "\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
