{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3efe82a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "AA  = ['A', 'R', 'N', 'D', 'C', 'E', 'Q', 'G', 'H', 'I','L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V']\n",
    "Amino_acid_dict = {'Ala': 'A',\n",
    "                   'Arg': 'R',\n",
    "                   'Asn': 'N',\n",
    "                   'Asp': 'D',\n",
    "                   'Cys': 'C',\n",
    "                   'Gln': 'Q',\n",
    "                   'Glu': 'E',\n",
    "                   'Gly': 'G',\n",
    "                   'His': 'H',\n",
    "                   'Ile': 'I',\n",
    "                   'Leu': 'L',\n",
    "                   'Lys': 'K',\n",
    "                   'Met': 'M',\n",
    "                   'Phe': 'F',\n",
    "                   'Pro': 'P',\n",
    "                   'Ser': 'S',\n",
    "                   'Thr': 'T',\n",
    "                   'Trp': 'W',\n",
    "                   'Tyr': 'Y',\n",
    "                   'Val': 'V',\n",
    "                   'Ter': '*'\n",
    "                  }\n",
    "def codon2aa(c, noq=False):              # Returns the amino acid character corresponding to the input codon.\n",
    "    if c[0]=='-' and c[1]=='-' and c[2]=='-': return '-'        # If all nucleotides are missing, return gap\n",
    "    elif c[0]=='-' or c[1]=='-' or c[2]=='-':                   # Else if some nucleotides are missing, return '?'\n",
    "        if noq: return '-'\n",
    "        else:   return '?'\n",
    "    # If the first or second nucleotide is ambiguous, AA cannot be determined, return 'X'\n",
    "    elif c[0] in ['W', 'S', 'M', 'K', 'R', 'Y'] or c[1] in ['W', 'S', 'M', 'K', 'R', 'Y']: return 'X'     \n",
    "                                                    \n",
    "    elif c[0]=='T':                                             # Else go to tree\n",
    "        if c[1]=='T':\n",
    "            if    c[2] in ['T', 'C', 'Y']: return 'F'\n",
    "            elif  c[2] in ['A', 'G', 'R']: return 'L'\n",
    "            else:                          return 'X'\n",
    "        elif c[1]=='C':                    return 'S'\n",
    "        elif c[1]=='A':\n",
    "            if    c[2] in ['T', 'C', 'Y']: return 'Y'\n",
    "            elif  c[2] in ['A', 'G', 'R']: return '*'\n",
    "            else:                          return 'X'\n",
    "        elif c[1]=='G':\n",
    "            if    c[2] in ['T', 'C', 'Y']: return 'C'\n",
    "            elif  c[2]=='A':               return '*'\n",
    "            elif  c[2]=='G':               return 'W'\n",
    "            else:                          return 'X'\n",
    "        else:                              return 'X'\n",
    "        \n",
    "    elif c[0]=='C':\n",
    "        if   c[1]=='T':                    return 'L'\n",
    "        elif c[1]=='C':                    return 'P'\n",
    "        elif c[1]=='A':\n",
    "            if    c[2] in ['T', 'C', 'Y']: return 'H'\n",
    "            elif  c[2] in ['A', 'G', 'R']: return 'Q'\n",
    "            else:                          return 'X'\n",
    "        elif c[1]=='G':                    return 'R'\n",
    "        else:                              return 'X'\n",
    "        \n",
    "    elif c[0]=='A':\n",
    "        if c[1]=='T':\n",
    "            if    c[2] in ['T', 'C', 'Y']: return 'I'\n",
    "            elif  c[2] in ['A', 'M', 'W']: return 'I'\n",
    "            elif  c[2]=='G':               return 'M'\n",
    "            else:                          return 'X'\n",
    "        elif c[1]=='C':                    return 'T'\n",
    "        elif c[1]=='A':\n",
    "            if    c[2] in ['T', 'C', 'Y']: return 'N'\n",
    "            elif  c[2] in ['A', 'G', 'R']: return 'K'\n",
    "            else:                          return 'X'\n",
    "        elif c[1]=='G':\n",
    "            if    c[2] in ['T', 'C', 'Y']: return 'S'\n",
    "            elif  c[2] in ['A', 'G', 'R']: return 'R'\n",
    "            else:                          return 'X'\n",
    "        else:                              return 'X'\n",
    "        \n",
    "    elif c[0]=='G':\n",
    "        if   c[1]=='T':                    return 'V'\n",
    "        elif c[1]=='C':                    return 'A'\n",
    "        elif c[1]=='A':\n",
    "            if    c[2] in ['T', 'C', 'Y']: return 'D'\n",
    "            elif  c[2] in ['A', 'G', 'R']: return 'E'\n",
    "            else:                          return 'X'\n",
    "        elif c[1]=='G':                    return 'G'\n",
    "        else:                              return 'X'\n",
    "\n",
    "    else:                                  return 'X'\n",
    "\n",
    "\n",
    "\n",
    "PREF_MERGED_DIR = './output/merged_preference/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f469c1c2",
   "metadata": {},
   "source": [
    "# HIV Env BF520 human host, rhesus host"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "01b1c9fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transform_host(index_file, pref_file, rep_num):\n",
    "    df_index = pd.read_csv(index_file, delimiter = ',')\n",
    "    df = pd.read_csv(pref_file)\n",
    "    df = df.dropna()\n",
    "    df['site'] = df['site'].astype('string')\n",
    "    df_index['original'] = df_index['original'].astype('string')\n",
    "    df_index['new'] = df_index['new'].astype('string')\n",
    "\n",
    "    df['site'] = df['site'].map(df_index.set_index('new')['original'])\n",
    "    df = df.dropna()\n",
    "    df['site'] = df['site'].astype('int')\n",
    "    \n",
    "    site_list = []\n",
    "    AA_list = []\n",
    "    df_pref_only = df[AA]\n",
    "    for i in df['site'].tolist():\n",
    "        site_list += [i]*20\n",
    "        AA_list += AA\n",
    "    df_pref = pd.DataFrame(columns=['site', 'amino_acid', rep_num]) \n",
    "    df_pref['site'] = site_list\n",
    "    df_pref['amino_acid'] = AA_list\n",
    "    df_pref[rep_num] = df_pref_only.values.flatten()\n",
    "    return df_pref\n",
    "\n",
    "# BG505\n",
    "index_file = './data/prefs/BG505_to_HXB2_numbering.txt'\n",
    "Target_name = 'HIV BG505'\n",
    "\n",
    "pref_file = './data/prefs/HIV Env BG505-1_prefs.csv'\n",
    "rep_num = 'rep_1'\n",
    "df1 = transform_host(index_file, pref_file, rep_num)\n",
    "\n",
    "pref_file = './data/prefs/HIV Env BG505-2_prefs.csv'\n",
    "rep_num = 'rep_2'\n",
    "df2 = transform_host(index_file, pref_file, rep_num)\n",
    "\n",
    "pref_file = './data/prefs/HIV Env BG505-3_prefs.csv'\n",
    "rep_num = 'rep_3'\n",
    "df3 = transform_host(index_file, pref_file, rep_num)\n",
    "\n",
    "df_merge = pd.merge(df1, df2, on = ['site', 'amino_acid'])\n",
    "df_merge = pd.merge(df_merge, df3, on = ['site', 'amino_acid'])\n",
    "df_merge['average'] = df_merge[['rep_1', 'rep_2', 'rep_3']].mean(axis=1)\n",
    "df_merge = df_merge.sort_values('site')\n",
    "df_merge.to_csv(PREF_MERGED_DIR + Target_name + '.csv.gz', index = False, compression = 'gzip')\n",
    "\n",
    "\n",
    "# BF520\n",
    "index_file = './data/prefs/BF520c2_to_HXB2.csv'\n",
    "Target_name = 'HIV BF520'\n",
    "\n",
    "pref_file = './data/prefs/HIV Env BF520-1_prefs.csv'\n",
    "rep_num = 'rep_1'\n",
    "df1 = transform_host(index_file, pref_file, rep_num)\n",
    "\n",
    "pref_file = './data/prefs/HIV Env BF520-2_prefs.csv'\n",
    "rep_num = 'rep_2'\n",
    "df2 = transform_host(index_file, pref_file, rep_num)\n",
    "\n",
    "pref_file = './data/prefs/HIV Env BF520-3_prefs.csv'\n",
    "rep_num = 'rep_3'\n",
    "df3 = transform_host(index_file, pref_file, rep_num)\n",
    "\n",
    "df_merge = pd.merge(df1, df2, on = ['site', 'amino_acid'])\n",
    "df_merge = pd.merge(df_merge, df3, on = ['site', 'amino_acid'])\n",
    "df_merge['average'] = df_merge[['rep_1', 'rep_2', 'rep_3']].mean(axis=1)\n",
    "df_merge = df_merge.sort_values('site')\n",
    "df_merge.to_csv(PREF_MERGED_DIR + Target_name + '.csv.gz', index = False, compression = 'gzip')\n",
    "\n",
    "# BF520 human host\n",
    "index_file = './data/prefs/BF520c2_to_HXB2.csv'\n",
    "Target_name = 'HIV BF520 human host'\n",
    "\n",
    "pref_file = './data/prefs/HIV BF520 human host-1_prefs.csv'\n",
    "rep_num = 'rep_1'\n",
    "df1 = transform_host(index_file, pref_file, rep_num)\n",
    "\n",
    "pref_file = './data/prefs/HIV BF520 human host-2_prefs.csv'\n",
    "rep_num = 'rep_2'\n",
    "df2 = transform_host(index_file, pref_file, rep_num)\n",
    "\n",
    "df_merge = pd.merge(df1, df2, on = ['site', 'amino_acid'])\n",
    "df_merge['average'] = df_merge[['rep_1', 'rep_2']].mean(axis=1)\n",
    "df_merge = df_merge.sort_values('site')\n",
    "df_merge.to_csv(PREF_MERGED_DIR + Target_name + '.csv.gz', index = False, compression = 'gzip')\n",
    "\n",
    "# BF520 rhesus host\n",
    "index_file = './data/prefs/BF520c2_to_HXB2.csv'\n",
    "Target_name = 'HIV BF520 rhesus host'\n",
    "\n",
    "pref_file = './data/prefs/HIV BF520 rhesus host-1_prefs.csv'\n",
    "rep_num = 'rep_1'\n",
    "df1 = transform_host(index_file, pref_file, rep_num)\n",
    "\n",
    "pref_file = './data/prefs/HIV BF520 rhesus host-2_prefs.csv'\n",
    "rep_num = 'rep_2'\n",
    "df2 = transform_host(index_file, pref_file, rep_num)\n",
    "\n",
    "df_merge = pd.merge(df1, df2, on = ['site', 'amino_acid'])\n",
    "df_merge['average'] = df_merge[['rep_1', 'rep_2']].mean(axis=1)\n",
    "df_merge = df_merge.sort_values('site')\n",
    "df_merge.to_csv(PREF_MERGED_DIR + Target_name + '.csv.gz', index = False, compression = 'gzip')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c39c1fa",
   "metadata": {},
   "source": [
    "# PR8, Aichi68C, MS, MxA, FP16, FP20, VRC34"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "751211f3",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def preference_merge(REP, Target_name, count_file):\n",
    "    pref_list = ['./data/prefs/'+Target_name+'-1_prefs.csv',\n",
    "                 './data/prefs/'+Target_name+'-2_prefs.csv',\n",
    "                 './data/prefs/'+Target_name+'-3_prefs.csv']\n",
    "\n",
    "    df_count = pd.read_csv(count_file)\n",
    "    site_list = df_count['site'].tolist()\n",
    "    site_list_merged = []\n",
    "    for site in site_list:\n",
    "        site_list_merged += [site]*20\n",
    "\n",
    "    aa_list_merged = []\n",
    "    for i in range(len(site_list)):\n",
    "        aa_list_merged += AA\n",
    "\n",
    "    rep=[i+1 for i in range(REP)]\n",
    "    rep_list=[]\n",
    "    for replicate in rep:\n",
    "        rep_list.append('rep_'+str(replicate))\n",
    "    df_pref_merged = pd.DataFrame(columns=['site', 'amino_acid'] + rep_list + ['average'])\n",
    "    df_pref_merged['site'] = site_list_merged\n",
    "    df_pref_merged['amino_acid'] = aa_list_merged\n",
    "\n",
    "    for i in rep:\n",
    "        df_pref = pd.read_csv(pref_list[i-1])\n",
    "        df_pref_aa = df_pref[AA]\n",
    "        df_pref_merged['rep_'+str(i)] = df_pref_aa.values.flatten()\n",
    "    df_pref_merged['average'] = df_pref_merged[rep_list].mean(axis = 1)\n",
    "    df_pref_merged.to_csv(PREF_MERGED_DIR + Target_name + '.csv.gz', index = False, compression = 'gzip')\n",
    "\n",
    "# PR8\n",
    "REP=3\n",
    "Target_name = 'PR8'\n",
    "count_file = './data/raw_data/PR8_DNA_codoncounts.csv'\n",
    "preference_merge(REP, Target_name, count_file)\n",
    "\n",
    "# Aichi68C\n",
    "REP=2\n",
    "Target_name = 'Aichi68C'\n",
    "count_file = './data/raw_data/Aichi68C_DNA_codoncounts.csv'\n",
    "preference_merge(REP, Target_name, count_file)\n",
    "\n",
    "# MS\n",
    "REP=2\n",
    "Target_name = 'MS'\n",
    "count_file = './data/raw_data/MS_DNA_codoncounts.csv'\n",
    "preference_merge(REP, Target_name, count_file)\n",
    "\n",
    "# MxA\n",
    "REP=2\n",
    "Target_name = 'MxA'\n",
    "count_file = './data/raw_data/MxA_DNA_codoncounts.csv'\n",
    "preference_merge(REP, Target_name, count_file)\n",
    "\n",
    "# MxAneg\n",
    "REP=2\n",
    "Target_name = 'MxAneg'\n",
    "count_file = './data/raw_data/MxAneg_DNA_codoncounts.csv'\n",
    "preference_merge(REP, Target_name, count_file)\n",
    "\n",
    "# FP16\n",
    "REP=2\n",
    "Target_name = 'HIV bnAbs FP16'\n",
    "count_file = './data/raw_data/FP16_DNA_codoncounts.csv'\n",
    "preference_merge(REP, Target_name, count_file)\n",
    "\n",
    "# FP20\n",
    "REP=2\n",
    "Target_name = 'HIV bnAbs FP20'\n",
    "count_file = './data/raw_data/FP20_DNA_codoncounts.csv'\n",
    "preference_merge(REP, Target_name, count_file)\n",
    "\n",
    "# VRC34\n",
    "REP=2\n",
    "Target_name = 'HIV bnAbs VRC34'\n",
    "count_file = './data/raw_data/VRC34_DNA_codoncounts.csv'\n",
    "preference_merge(REP, Target_name, count_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72366f9d",
   "metadata": {},
   "source": [
    "# Flu WSN, A549, CCL141, Matrix_M1, ZIKV, Perth2009"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "33ffd48e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transform_WSN(df, rep_num, WSN):\n",
    "    if WSN == True:\n",
    "        df = df.drop(['WT_AA', 'SITE_ENTROPY'], axis = 1)\n",
    "        df.set_axis(['site']+[i[3] for i in df.columns[1:]], axis=1,inplace=True)\n",
    "    site_list = []\n",
    "    AA_list = []\n",
    "    df_pref_only = df[AA]\n",
    "    for i in df['site'].tolist():\n",
    "        site_list += [i]*20\n",
    "        AA_list += AA\n",
    "    df_pref = pd.DataFrame(columns=['site', 'amino_acid', rep_num]) \n",
    "    df_pref['site'] = site_list\n",
    "    df_pref['amino_acid'] = AA_list\n",
    "    df_pref[rep_num] = df_pref_only.values.flatten()\n",
    "    return df_pref\n",
    "\n",
    "# WSN\n",
    "Target_name = 'WSN'\n",
    "df = pd.read_csv('./data/prefs/WSN-1_prefs.txt', delimiter = '\\t')\n",
    "rep_num = 'rep_1'\n",
    "df1 = transform_WSN(df, rep_num, True)\n",
    "\n",
    "df = pd.read_csv('./data/prefs/WSN-2_prefs.txt', delimiter = '\\t')\n",
    "rep_num = 'rep_2'\n",
    "df2 = transform_WSN(df, rep_num, True)\n",
    "\n",
    "df = pd.read_csv('./data/prefs/WSN-3_prefs.txt', delimiter = '\\t')\n",
    "rep_num = 'rep_3'\n",
    "df3 = transform_WSN(df, rep_num, True)\n",
    "\n",
    "df_merge = pd.merge(df1, df2, on = ['site', 'amino_acid'])\n",
    "df_merge = pd.merge(df_merge, df3, on = ['site', 'amino_acid'])\n",
    "df_merge['average'] = df_merge[['rep_1', 'rep_2', 'rep_3']].mean(axis=1)\n",
    "df_merge = df_merge.sort_values('site')\n",
    "df_merge.to_csv(PREF_MERGED_DIR + Target_name + '.csv.gz', index = False, compression = 'gzip')\n",
    "\n",
    "# A549\n",
    "Target_name = 'A549'\n",
    "df = pd.read_csv('./data/prefs/A549-1_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_1'\n",
    "df1 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df = pd.read_csv('./data/prefs/A549-2_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_2'\n",
    "df2 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df_merge = pd.merge(df1, df2, on = ['site', 'amino_acid'])\n",
    "df_merge['average'] = df_merge[['rep_1', 'rep_2']].mean(axis=1)\n",
    "df_merge = df_merge.sort_values('site')\n",
    "df_merge.to_csv(PREF_MERGED_DIR + Target_name + '.csv.gz', index = False, compression = 'gzip')\n",
    "\n",
    "#CCL141\n",
    "Target_name = 'CCL141'\n",
    "df = pd.read_csv('./data/prefs/CCL141-1_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_1'\n",
    "df1 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df = pd.read_csv('./data/prefs/CCL141-2_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_2'\n",
    "df2 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df = pd.read_csv('./data/prefs/CCL141-3_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_3'\n",
    "df3 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df_merge = pd.merge(df1, df2, on = ['site', 'amino_acid'])\n",
    "df_merge = pd.merge(df_merge, df3, on = ['site', 'amino_acid'])\n",
    "df_merge['average'] = df_merge[['rep_1', 'rep_2', 'rep_3']].mean(axis=1)\n",
    "df_merge = df_merge.sort_values('site')\n",
    "df_merge.to_csv(PREF_MERGED_DIR + Target_name + '.csv.gz', index = False, compression = 'gzip')\n",
    "\n",
    "# Matrix_M1\n",
    "Target_name = 'Matrix_M1'\n",
    "df = pd.read_csv('./data/prefs/Matrix_M1-1_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_1'\n",
    "df1 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df = pd.read_csv('./data/prefs/Matrix_M1-2_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_2'\n",
    "df2 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df = pd.read_csv('./data/prefs/Matrix_M1-3_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_3'\n",
    "df3 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df_merge = pd.merge(df1, df2, on = ['site', 'amino_acid'])\n",
    "df_merge = pd.merge(df_merge, df3, on = ['site', 'amino_acid'])\n",
    "df_merge['average'] = df_merge[['rep_1', 'rep_2', 'rep_3']].mean(axis=1)\n",
    "df_merge = df_merge.sort_values('site')\n",
    "df_merge.to_csv(PREF_MERGED_DIR + Target_name + '.csv.gz', index = False, compression = 'gzip')\n",
    "\n",
    "# ZIKV\n",
    "Target_name = 'ZIKV'\n",
    "df = pd.read_csv('./data/prefs/ZIKV-1_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_1'\n",
    "df1 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df = pd.read_csv('./data/prefs/ZIKV-2_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_2'\n",
    "df2 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df = pd.read_csv('./data/prefs/ZIKV-3_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_3'\n",
    "df3 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df_merge = pd.merge(df1, df2, on = ['site', 'amino_acid'])\n",
    "df_merge = pd.merge(df_merge, df3, on = ['site', 'amino_acid'])\n",
    "df_merge['average'] = df_merge[['rep_1', 'rep_2', 'rep_3']].mean(axis=1)\n",
    "df_merge = df_merge.sort_values('site')\n",
    "df_merge.to_csv(PREF_MERGED_DIR + Target_name + '.csv.gz', index = False, compression = 'gzip')\n",
    "\n",
    "\n",
    "# Perth2009\n",
    "Target_name = 'Perth2009'\n",
    "df = pd.read_csv('./data/prefs/Perth2009-1_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_1'\n",
    "df1 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df = pd.read_csv('./data/prefs/Perth2009-2_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_2'\n",
    "df2 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df = pd.read_csv('./data/prefs/Perth2009-3_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_3'\n",
    "df3 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df = pd.read_csv('./data/prefs/Perth2009-4_prefs.csv', delimiter = ',')\n",
    "rep_num = 'rep_4'\n",
    "df4 = transform_WSN(df, rep_num, False)\n",
    "\n",
    "df_merge = pd.merge(df1, df2, on = ['site', 'amino_acid'])\n",
    "df_merge = pd.merge(df_merge, df3, on = ['site', 'amino_acid'])\n",
    "df_merge = pd.merge(df_merge, df4, on = ['site', 'amino_acid'])\n",
    "df_merge['average'] = df_merge[['rep_1', 'rep_2', 'rep_3', 'rep_4']].mean(axis=1)\n",
    "df_merge = df_merge.sort_values('site')\n",
    "df_merge.to_csv(PREF_MERGED_DIR + Target_name + '.csv.gz', index = False, compression = 'gzip')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd83fae1",
   "metadata": {},
   "source": [
    "# TpoR, TpoR_S505N, Ube4b, YAP1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bcb72006",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Transfor_MaveDB(Target_name, pref_column, pref_file):\n",
    "    df = pd.read_csv(pref_file)\n",
    "    site_list = []\n",
    "    amino_acid = []\n",
    "    pref_list = []\n",
    "    for _ in range(len(pref_column)):\n",
    "        pref_list.append([])\n",
    "    for i in range(df.shape[0]):\n",
    "\n",
    "        parser = re.split('(\\d+)', df['hgvs_pro'][i])\n",
    "        if len(parser) == 3 and '?' not in parser:\n",
    "            site_list.append(int(parser[1]))\n",
    "            amino_acid.append(Amino_acid_dict[parser[-1]])\n",
    "            for idx in range(len(pref_column)):\n",
    "                pref_list[idx].append(df[pref_column[idx]][i])\n",
    "\n",
    "    df_merge = pd.DataFrame(columns=['site', 'amino_acid']+['rep_'+str(i+1) for i in range(len(pref_column))]+['average'])\n",
    "    df_merge['site'] = site_list\n",
    "    df_merge['amino_acid']= amino_acid\n",
    "    for i in range(len(pref_column)):\n",
    "        df_merge['rep_'+str(i+1)] = pref_list[i]\n",
    "    df_merge['average'] = df_merge[['rep_'+str(i+1) for i in range(len(pref_column))]].mean(axis=1)\n",
    "    df_merge = df_merge.sort_values('site')\n",
    "    df_merge.to_csv(PREF_MERGED_DIR + Target_name + '.csv.gz', index = False, compression = 'gzip')\n",
    "    \n",
    "\n",
    "Target_name = 'TpoR'\n",
    "pref_file = './data/prefs/TpoR_prefs.csv.gz'\n",
    "pref_column = ['score_Replicate_A', \n",
    "               'score_Replicate_B', \n",
    "               'score_Replicate_C', \n",
    "               'score_Replicate_D', \n",
    "               'score_Replicate_E', \n",
    "               'score_Replicate_F.1']\n",
    "Transfor_MaveDB(Target_name, pref_column, pref_file)\n",
    "\n",
    "\n",
    "Target_name = 'TpoR_S505N'\n",
    "pref_file = './data/prefs/TpoR_S505N_prefs.csv.gz'\n",
    "pref_column = ['score_Replicate_1', \n",
    "               'score_Replicate_2', \n",
    "               'score_Replicate_3', \n",
    "               'score_Replicate_4', \n",
    "               'score_Replicate_5', \n",
    "               'score_Replicate_6.1']\n",
    "Transfor_MaveDB(Target_name, pref_column, pref_file)\n",
    "\n",
    "\n",
    "Target_name = 'Ube4b'\n",
    "pref_file = './data/prefs/Ube4b_prefs.csv.gz'\n",
    "pref_column = ['score_Rep_2', \n",
    "               'score_Rep_3']\n",
    "Transfor_MaveDB(Target_name, pref_column, pref_file)\n",
    "\n",
    "\n",
    "Target_name = 'YAP1'\n",
    "pref_file = './data/prefs/YAP1_prefs.csv.gz'\n",
    "pref_column = ['score_101208', \n",
    "               'score_110307']\n",
    "Transfor_MaveDB(Target_name, pref_column, pref_file)\n",
    "\n",
    "Target_name = 'E3'\n",
    "pref_file = './data/prefs/E3_prefs.csv'\n",
    "pref_column = ['score_PlusE2NewRep3', \n",
    "               'score_PlusE2NewRep4',\n",
    "               'score_PlusE2NewRep5',\n",
    "               'score_PlusE2Rep3', \n",
    "               'score_PlusE2Rep4',\n",
    "               'score_PlusE2Rep5']\n",
    "Transfor_MaveDB(Target_name, pref_column, pref_file)\n",
    "\n",
    "Target_name = 'Y2H_1'\n",
    "pref_file = './data/prefs/Y2H_prefs.csv'\n",
    "pref_column = ['score_Y2H_1_Rep1', \n",
    "               'score_Y2H_1_Rep2',\n",
    "               'score_Y2H_1_Rep3']\n",
    "Transfor_MaveDB(Target_name, pref_column, pref_file)\n",
    "\n",
    "Target_name = 'Y2H_2'\n",
    "pref_file = './data/prefs/Y2H_prefs.csv'\n",
    "pref_column = ['score_Y2H_2_Rep1', \n",
    "               'score_Y2H_2_Rep2',\n",
    "               'score_Y2H_2_Rep3']\n",
    "Transfor_MaveDB(Target_name, pref_column, pref_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f0d6734",
   "metadata": {},
   "source": [
    "# DBR1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7c71acd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('./data/prefs/DBR1_prefs.csv.gz')\n",
    "df = df.dropna()\n",
    "df = df.drop(['Sequence', 'Reference AA', 'Variant Class'], axis = 1)\n",
    "df = df.rename(columns={'Affected codon': 'site', \n",
    "                        'Substituted AA': 'amino_acid',\n",
    "                        'Day 11 log2 enrichment score (replicate 1)': 'rep_1', \n",
    "                        'Day 11 log2 enrichment score (replicate 2)': 'rep_2'})\n",
    "df['average'] = df[['rep_1', 'rep_2']].mean(axis = 1)\n",
    "df['site'] = df['site'].astype('int')\n",
    "df_merge = df.sort_values('site')\n",
    "df_merge.to_csv(PREF_MERGED_DIR+'DBR1.csv.gz', index = False, compression = 'gzip')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbb85336",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfb28438",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "36474ec5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>rep_3</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>rep_3_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.009139</td>\n",
       "      <td>-1.203026e-02</td>\n",
       "      <td>-8.203877e-03</td>\n",
       "      <td>-0.025103</td>\n",
       "      <td>-0.226252</td>\n",
       "      <td>-0.236142</td>\n",
       "      <td>-0.222815</td>\n",
       "      <td>-0.487661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.037421</td>\n",
       "      <td>-3.088728e-02</td>\n",
       "      <td>-2.985570e-02</td>\n",
       "      <td>-0.081780</td>\n",
       "      <td>-0.254535</td>\n",
       "      <td>-0.254999</td>\n",
       "      <td>-0.244466</td>\n",
       "      <td>-0.544338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.011831</td>\n",
       "      <td>-9.553025e-03</td>\n",
       "      <td>-1.172133e-02</td>\n",
       "      <td>-0.028050</td>\n",
       "      <td>-0.228944</td>\n",
       "      <td>-0.233665</td>\n",
       "      <td>-0.226332</td>\n",
       "      <td>-0.490608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.003844</td>\n",
       "      <td>-3.454097e-03</td>\n",
       "      <td>-3.031401e-03</td>\n",
       "      <td>-0.008886</td>\n",
       "      <td>-0.220957</td>\n",
       "      <td>-0.227566</td>\n",
       "      <td>-0.217642</td>\n",
       "      <td>-0.471444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.006201</td>\n",
       "      <td>-9.180511e-03</td>\n",
       "      <td>-3.895432e-03</td>\n",
       "      <td>-0.016547</td>\n",
       "      <td>-0.223314</td>\n",
       "      <td>-0.233292</td>\n",
       "      <td>-0.218506</td>\n",
       "      <td>-0.479104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10579</th>\n",
       "      <td>504</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.017051</td>\n",
       "      <td>-1.609082e-02</td>\n",
       "      <td>-1.713299e-02</td>\n",
       "      <td>-0.041580</td>\n",
       "      <td>-0.189193</td>\n",
       "      <td>-0.157002</td>\n",
       "      <td>-0.174478</td>\n",
       "      <td>-0.435361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10580</th>\n",
       "      <td>504</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.001415</td>\n",
       "      <td>-1.735121e-07</td>\n",
       "      <td>-1.789024e-03</td>\n",
       "      <td>-0.002801</td>\n",
       "      <td>-0.173557</td>\n",
       "      <td>-0.140912</td>\n",
       "      <td>-0.159134</td>\n",
       "      <td>-0.396582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10581</th>\n",
       "      <td>504</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.001415</td>\n",
       "      <td>-1.812894e-03</td>\n",
       "      <td>-2.320765e-07</td>\n",
       "      <td>-0.002808</td>\n",
       "      <td>-0.173557</td>\n",
       "      <td>-0.142724</td>\n",
       "      <td>-0.157346</td>\n",
       "      <td>-0.396589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10582</th>\n",
       "      <td>504</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.003488</td>\n",
       "      <td>1.104854e-03</td>\n",
       "      <td>3.992824e-03</td>\n",
       "      <td>0.006708</td>\n",
       "      <td>-0.175630</td>\n",
       "      <td>-0.139806</td>\n",
       "      <td>-0.153353</td>\n",
       "      <td>-0.387073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10583</th>\n",
       "      <td>504</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.036092</td>\n",
       "      <td>-2.152075e-02</td>\n",
       "      <td>-2.298831e-02</td>\n",
       "      <td>-0.068429</td>\n",
       "      <td>-0.208234</td>\n",
       "      <td>-0.162432</td>\n",
       "      <td>-0.180334</td>\n",
       "      <td>-0.462210</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10584 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator     rep_1         rep_2         rep_3  \\\n",
       "0         1          A         False -0.009139 -1.203026e-02 -8.203877e-03   \n",
       "1         1          R         False -0.037421 -3.088728e-02 -2.985570e-02   \n",
       "2         1          N         False -0.011831 -9.553025e-03 -1.172133e-02   \n",
       "3         1          D         False -0.003844 -3.454097e-03 -3.031401e-03   \n",
       "4         1          C         False -0.006201 -9.180511e-03 -3.895432e-03   \n",
       "...     ...        ...           ...       ...           ...           ...   \n",
       "10579   504          T         False -0.017051 -1.609082e-02 -1.713299e-02   \n",
       "10580   504          W         False -0.001415 -1.735121e-07 -1.789024e-03   \n",
       "10581   504          Y         False -0.001415 -1.812894e-03 -2.320765e-07   \n",
       "10582   504          V         False -0.003488  1.104854e-03  3.992824e-03   \n",
       "10583   504          *         False -0.036092 -2.152075e-02 -2.298831e-02   \n",
       "\n",
       "          joint  rep_1_gauged  rep_2_gauged  rep_3_gauged  joint_gauged  \n",
       "0     -0.025103     -0.226252     -0.236142     -0.222815     -0.487661  \n",
       "1     -0.081780     -0.254535     -0.254999     -0.244466     -0.544338  \n",
       "2     -0.028050     -0.228944     -0.233665     -0.226332     -0.490608  \n",
       "3     -0.008886     -0.220957     -0.227566     -0.217642     -0.471444  \n",
       "4     -0.016547     -0.223314     -0.233292     -0.218506     -0.479104  \n",
       "...         ...           ...           ...           ...           ...  \n",
       "10579 -0.041580     -0.189193     -0.157002     -0.174478     -0.435361  \n",
       "10580 -0.002801     -0.173557     -0.140912     -0.159134     -0.396582  \n",
       "10581 -0.002808     -0.173557     -0.142724     -0.157346     -0.396589  \n",
       "10582  0.006708     -0.175630     -0.139806     -0.153353     -0.387073  \n",
       "10583 -0.068429     -0.208234     -0.162432     -0.180334     -0.462210  \n",
       "\n",
       "[10584 rows x 11 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def wildtype(codon_file, selection_file, output_file, replicates):\n",
    "    df = pd.read_csv(codon_file)\n",
    "    WT_aa = [codon2aa(i) for i in df['wildtype'].tolist()]\n",
    "    df['WT_aa'] = WT_aa\n",
    "    df['WT_indicator'] = True\n",
    "    df_WT = df[['site', 'WT_aa','WT_indicator']]\n",
    "    df_WT = df_WT.rename(columns={'WT_aa': 'amino_acid'})\n",
    "    df_selection = pd.read_csv(selection_file)\n",
    "    \n",
    "    if 'gauged' not in ' '.join(df_selection.columns.tolist()):\n",
    "        df_merge = pd.merge(df_WT, df_selection, on=['site','amino_acid'], how='right')\n",
    "        df_merge['WT_indicator'] = df_merge['WT_indicator'].fillna(False)\n",
    "        column_name = df_merge.columns[-(replicates+1):].tolist()\n",
    "        gauged_name = [i+'_gauged' for i in column_name]\n",
    "        gauged_list = []\n",
    "        for i in gauged_name:\n",
    "            df_merge[i] = 0\n",
    "            gauged_list.append([])\n",
    "\n",
    "        for site in df_merge['site'].unique():\n",
    "            for i in range(len(column_name)):\n",
    "                gauge_value = df_merge[(df_merge['site']==site)&(df_merge['WT_indicator'])][column_name[i]].values[0]\n",
    "                original_list = df_merge[df_merge['site']==site][column_name[i]].tolist()\n",
    "#                 print(gauge_value)\n",
    "                gauged_list[i] +=  [i - gauge_value for i in original_list]\n",
    "\n",
    "        for i in range(len(gauged_name)):\n",
    "            df_merge[gauged_name[i]] = gauged_list[i]\n",
    "\n",
    "        df_merge.to_csv(output_file, compression = 'gzip', index = False)\n",
    "        print('gauge done')\n",
    "        return df_merge\n",
    "    else:\n",
    "        print('gauge existed')\n",
    "\n",
    "\n",
    "codon_file = './data/raw_data/ZIKV_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/ZIKV.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/ZIKV.csv.gz'\n",
    "replicates = 3\n",
    "wildtype(codon_file, selection_file, output_file, replicates)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6352682c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>rep_3</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>rep_3_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.015366</td>\n",
       "      <td>-0.042260</td>\n",
       "      <td>0.012669</td>\n",
       "      <td>0.015350</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.021677</td>\n",
       "      <td>-0.006807</td>\n",
       "      <td>-0.002920</td>\n",
       "      <td>-0.015414</td>\n",
       "      <td>-0.006311</td>\n",
       "      <td>0.035453</td>\n",
       "      <td>-0.015589</td>\n",
       "      <td>-0.030764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>30</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.021327</td>\n",
       "      <td>-0.016567</td>\n",
       "      <td>-0.014036</td>\n",
       "      <td>-0.040368</td>\n",
       "      <td>-0.005960</td>\n",
       "      <td>0.025694</td>\n",
       "      <td>-0.026705</td>\n",
       "      <td>-0.055718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.022502</td>\n",
       "      <td>0.024385</td>\n",
       "      <td>0.013555</td>\n",
       "      <td>0.020483</td>\n",
       "      <td>-0.007135</td>\n",
       "      <td>0.066645</td>\n",
       "      <td>0.000886</td>\n",
       "      <td>0.005133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>30</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.010532</td>\n",
       "      <td>-0.018857</td>\n",
       "      <td>-0.020996</td>\n",
       "      <td>-0.042158</td>\n",
       "      <td>0.004834</td>\n",
       "      <td>0.023403</td>\n",
       "      <td>-0.033665</td>\n",
       "      <td>-0.057508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14065</th>\n",
       "      <td>699</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.049961</td>\n",
       "      <td>0.011372</td>\n",
       "      <td>0.004885</td>\n",
       "      <td>-0.020309</td>\n",
       "      <td>-0.127939</td>\n",
       "      <td>0.080636</td>\n",
       "      <td>-0.040203</td>\n",
       "      <td>-0.069744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14066</th>\n",
       "      <td>699</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.017808</td>\n",
       "      <td>0.002355</td>\n",
       "      <td>-0.013677</td>\n",
       "      <td>-0.026750</td>\n",
       "      <td>-0.095787</td>\n",
       "      <td>0.071619</td>\n",
       "      <td>-0.058764</td>\n",
       "      <td>-0.076185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14067</th>\n",
       "      <td>699</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>0.011237</td>\n",
       "      <td>-0.018501</td>\n",
       "      <td>-0.023079</td>\n",
       "      <td>-0.028164</td>\n",
       "      <td>-0.066741</td>\n",
       "      <td>0.050763</td>\n",
       "      <td>-0.068166</td>\n",
       "      <td>-0.077599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14068</th>\n",
       "      <td>699</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.028438</td>\n",
       "      <td>0.050798</td>\n",
       "      <td>-0.015883</td>\n",
       "      <td>0.011675</td>\n",
       "      <td>-0.106416</td>\n",
       "      <td>0.120062</td>\n",
       "      <td>-0.060970</td>\n",
       "      <td>-0.037761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14069</th>\n",
       "      <td>699</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.062179</td>\n",
       "      <td>-0.031553</td>\n",
       "      <td>-0.063012</td>\n",
       "      <td>-0.148194</td>\n",
       "      <td>-0.140157</td>\n",
       "      <td>0.037711</td>\n",
       "      <td>-0.108099</td>\n",
       "      <td>-0.197629</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14070 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator     rep_1     rep_2     rep_3     joint  \\\n",
       "0        30          A          True -0.015366 -0.042260  0.012669  0.015350   \n",
       "1        30          R         False -0.021677 -0.006807 -0.002920 -0.015414   \n",
       "2        30          N         False -0.021327 -0.016567 -0.014036 -0.040368   \n",
       "3        30          D         False -0.022502  0.024385  0.013555  0.020483   \n",
       "4        30          C         False -0.010532 -0.018857 -0.020996 -0.042158   \n",
       "...     ...        ...           ...       ...       ...       ...       ...   \n",
       "14065   699          T         False -0.049961  0.011372  0.004885 -0.020309   \n",
       "14066   699          W         False -0.017808  0.002355 -0.013677 -0.026750   \n",
       "14067   699          Y         False  0.011237 -0.018501 -0.023079 -0.028164   \n",
       "14068   699          V         False -0.028438  0.050798 -0.015883  0.011675   \n",
       "14069   699          *         False -0.062179 -0.031553 -0.063012 -0.148194   \n",
       "\n",
       "       rep_1_gauged  rep_2_gauged  rep_3_gauged  joint_gauged  \n",
       "0          0.000000      0.000000      0.000000      0.000000  \n",
       "1         -0.006311      0.035453     -0.015589     -0.030764  \n",
       "2         -0.005960      0.025694     -0.026705     -0.055718  \n",
       "3         -0.007135      0.066645      0.000886      0.005133  \n",
       "4          0.004834      0.023403     -0.033665     -0.057508  \n",
       "...             ...           ...           ...           ...  \n",
       "14065     -0.127939      0.080636     -0.040203     -0.069744  \n",
       "14066     -0.095787      0.071619     -0.058764     -0.076185  \n",
       "14067     -0.066741      0.050763     -0.068166     -0.077599  \n",
       "14068     -0.106416      0.120062     -0.060970     -0.037761  \n",
       "14069     -0.140157      0.037711     -0.108099     -0.197629  \n",
       "\n",
       "[14070 rows x 11 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/BG505_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/HIV Env BG505.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/HIV Env BG505.csv.gz'\n",
    "replicates = 3\n",
    "wildtype(codon_file, selection_file, output_file, replicates)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "404233f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>rep_3</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>rep_3_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>-1.438800e-01</td>\n",
       "      <td>-1.944390e-01</td>\n",
       "      <td>-2.441822e-01</td>\n",
       "      <td>-2.950637e-01</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.012725e-07</td>\n",
       "      <td>-2.262431e-07</td>\n",
       "      <td>-7.253101e-08</td>\n",
       "      <td>-4.000466e-07</td>\n",
       "      <td>0.143880</td>\n",
       "      <td>0.194439</td>\n",
       "      <td>0.244182</td>\n",
       "      <td>0.295063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>30</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.143880</td>\n",
       "      <td>0.194439</td>\n",
       "      <td>0.244182</td>\n",
       "      <td>0.295064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>-5.315121e-07</td>\n",
       "      <td>-3.976840e-07</td>\n",
       "      <td>-5.682080e-07</td>\n",
       "      <td>-1.497404e-06</td>\n",
       "      <td>0.143879</td>\n",
       "      <td>0.194439</td>\n",
       "      <td>0.244182</td>\n",
       "      <td>0.295062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>30</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.143880</td>\n",
       "      <td>0.194439</td>\n",
       "      <td>0.244182</td>\n",
       "      <td>0.295064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13897</th>\n",
       "      <td>691</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>2.789343e-03</td>\n",
       "      <td>1.615925e-02</td>\n",
       "      <td>3.356764e-02</td>\n",
       "      <td>4.721649e-02</td>\n",
       "      <td>-0.052368</td>\n",
       "      <td>0.031392</td>\n",
       "      <td>0.055965</td>\n",
       "      <td>0.029068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13898</th>\n",
       "      <td>691</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-3.017728e-03</td>\n",
       "      <td>-1.076386e-02</td>\n",
       "      <td>-4.209948e-03</td>\n",
       "      <td>-1.539347e-02</td>\n",
       "      <td>-0.058175</td>\n",
       "      <td>0.004469</td>\n",
       "      <td>0.018188</td>\n",
       "      <td>-0.033542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13899</th>\n",
       "      <td>691</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.249844e-02</td>\n",
       "      <td>-6.414506e-04</td>\n",
       "      <td>9.570118e-03</td>\n",
       "      <td>-2.401218e-03</td>\n",
       "      <td>-0.067656</td>\n",
       "      <td>0.014591</td>\n",
       "      <td>0.031968</td>\n",
       "      <td>-0.020550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13900</th>\n",
       "      <td>691</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>1.755803e-02</td>\n",
       "      <td>7.601935e-03</td>\n",
       "      <td>-1.405762e-02</td>\n",
       "      <td>1.332714e-02</td>\n",
       "      <td>-0.037600</td>\n",
       "      <td>0.022835</td>\n",
       "      <td>0.008340</td>\n",
       "      <td>-0.004821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13901</th>\n",
       "      <td>691</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>-9.526063e-02</td>\n",
       "      <td>-9.345170e-02</td>\n",
       "      <td>-5.698323e-02</td>\n",
       "      <td>-2.164138e-01</td>\n",
       "      <td>-0.150418</td>\n",
       "      <td>-0.078219</td>\n",
       "      <td>-0.034586</td>\n",
       "      <td>-0.234562</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13902 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator         rep_1         rep_2  \\\n",
       "0        30          A          True -1.438800e-01 -1.944390e-01   \n",
       "1        30          R         False -1.012725e-07 -2.262431e-07   \n",
       "2        30          N         False  0.000000e+00  0.000000e+00   \n",
       "3        30          D         False -5.315121e-07 -3.976840e-07   \n",
       "4        30          C         False  0.000000e+00  0.000000e+00   \n",
       "...     ...        ...           ...           ...           ...   \n",
       "13897   691          T         False  2.789343e-03  1.615925e-02   \n",
       "13898   691          W         False -3.017728e-03 -1.076386e-02   \n",
       "13899   691          Y         False -1.249844e-02 -6.414506e-04   \n",
       "13900   691          V         False  1.755803e-02  7.601935e-03   \n",
       "13901   691          *         False -9.526063e-02 -9.345170e-02   \n",
       "\n",
       "              rep_3         joint  rep_1_gauged  rep_2_gauged  rep_3_gauged  \\\n",
       "0     -2.441822e-01 -2.950637e-01      0.000000      0.000000      0.000000   \n",
       "1     -7.253101e-08 -4.000466e-07      0.143880      0.194439      0.244182   \n",
       "2      0.000000e+00  0.000000e+00      0.143880      0.194439      0.244182   \n",
       "3     -5.682080e-07 -1.497404e-06      0.143879      0.194439      0.244182   \n",
       "4      0.000000e+00  0.000000e+00      0.143880      0.194439      0.244182   \n",
       "...             ...           ...           ...           ...           ...   \n",
       "13897  3.356764e-02  4.721649e-02     -0.052368      0.031392      0.055965   \n",
       "13898 -4.209948e-03 -1.539347e-02     -0.058175      0.004469      0.018188   \n",
       "13899  9.570118e-03 -2.401218e-03     -0.067656      0.014591      0.031968   \n",
       "13900 -1.405762e-02  1.332714e-02     -0.037600      0.022835      0.008340   \n",
       "13901 -5.698323e-02 -2.164138e-01     -0.150418     -0.078219     -0.034586   \n",
       "\n",
       "       joint_gauged  \n",
       "0          0.000000  \n",
       "1          0.295063  \n",
       "2          0.295064  \n",
       "3          0.295062  \n",
       "4          0.295064  \n",
       "...             ...  \n",
       "13897      0.029068  \n",
       "13898     -0.033542  \n",
       "13899     -0.020550  \n",
       "13900     -0.004821  \n",
       "13901     -0.234562  \n",
       "\n",
       "[13902 rows x 11 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/BF520_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/HIV Env BF520.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/HIV Env BF520.csv.gz'\n",
    "replicates = 3\n",
    "wildtype(codon_file, selection_file, output_file, replicates)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7bd650a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>-1.252172e-01</td>\n",
       "      <td>-4.717869e-02</td>\n",
       "      <td>-1.171115e-01</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.367643e-07</td>\n",
       "      <td>-2.106226e-07</td>\n",
       "      <td>-3.473869e-07</td>\n",
       "      <td>0.125217</td>\n",
       "      <td>0.047178</td>\n",
       "      <td>0.117111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>30</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.125217</td>\n",
       "      <td>0.047179</td>\n",
       "      <td>0.117112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>-4.674272e-07</td>\n",
       "      <td>-5.533146e-07</td>\n",
       "      <td>-1.020742e-06</td>\n",
       "      <td>0.125217</td>\n",
       "      <td>0.047178</td>\n",
       "      <td>0.117111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>30</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.125217</td>\n",
       "      <td>0.047179</td>\n",
       "      <td>0.117112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14065</th>\n",
       "      <td>699</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>-8.224614e-09</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-8.224614e-09</td>\n",
       "      <td>-0.064921</td>\n",
       "      <td>-0.102062</td>\n",
       "      <td>-0.129689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14066</th>\n",
       "      <td>699</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-3.853823e-03</td>\n",
       "      <td>-2.121521e-06</td>\n",
       "      <td>-3.727067e-03</td>\n",
       "      <td>-0.068775</td>\n",
       "      <td>-0.102064</td>\n",
       "      <td>-0.133416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14067</th>\n",
       "      <td>699</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.057307e-06</td>\n",
       "      <td>-1.156271e-06</td>\n",
       "      <td>-2.213577e-06</td>\n",
       "      <td>-0.064922</td>\n",
       "      <td>-0.102063</td>\n",
       "      <td>-0.129691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14068</th>\n",
       "      <td>699</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-8.171748e-09</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-8.171748e-09</td>\n",
       "      <td>-0.064921</td>\n",
       "      <td>-0.102062</td>\n",
       "      <td>-0.129689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14069</th>\n",
       "      <td>699</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>-3.735946e-02</td>\n",
       "      <td>-1.195814e-01</td>\n",
       "      <td>-1.350577e-01</td>\n",
       "      <td>-0.102281</td>\n",
       "      <td>-0.221644</td>\n",
       "      <td>-0.264746</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14070 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator         rep_1         rep_2  \\\n",
       "0        30          A          True -1.252172e-01 -4.717869e-02   \n",
       "1        30          R         False -1.367643e-07 -2.106226e-07   \n",
       "2        30          N         False  0.000000e+00  0.000000e+00   \n",
       "3        30          D         False -4.674272e-07 -5.533146e-07   \n",
       "4        30          C         False  0.000000e+00  0.000000e+00   \n",
       "...     ...        ...           ...           ...           ...   \n",
       "14065   699          T         False -8.224614e-09  0.000000e+00   \n",
       "14066   699          W         False -3.853823e-03 -2.121521e-06   \n",
       "14067   699          Y         False -1.057307e-06 -1.156271e-06   \n",
       "14068   699          V         False -8.171748e-09  0.000000e+00   \n",
       "14069   699          *         False -3.735946e-02 -1.195814e-01   \n",
       "\n",
       "              joint  rep_1_gauged  rep_2_gauged  joint_gauged  \n",
       "0     -1.171115e-01      0.000000      0.000000      0.000000  \n",
       "1     -3.473869e-07      0.125217      0.047178      0.117111  \n",
       "2      0.000000e+00      0.125217      0.047179      0.117112  \n",
       "3     -1.020742e-06      0.125217      0.047178      0.117111  \n",
       "4      0.000000e+00      0.125217      0.047179      0.117112  \n",
       "...             ...           ...           ...           ...  \n",
       "14065 -8.224614e-09     -0.064921     -0.102062     -0.129689  \n",
       "14066 -3.727067e-03     -0.068775     -0.102064     -0.133416  \n",
       "14067 -2.213577e-06     -0.064922     -0.102063     -0.129691  \n",
       "14068 -8.171748e-09     -0.064921     -0.102062     -0.129689  \n",
       "14069 -1.350577e-01     -0.102281     -0.221644     -0.264746  \n",
       "\n",
       "[14070 rows x 9 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/human_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/HIV BF520 human host.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/HIV BF520 human host.csv.gz'\n",
    "replicates = 2\n",
    "wildtype(codon_file, selection_file, output_file, replicates)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c704c21c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>-4.105970e-02</td>\n",
       "      <td>-5.908830e-03</td>\n",
       "      <td>-4.336264e-02</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.367643e-08</td>\n",
       "      <td>-2.106226e-08</td>\n",
       "      <td>-3.473869e-08</td>\n",
       "      <td>0.041060</td>\n",
       "      <td>0.005909</td>\n",
       "      <td>0.043363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>30</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.041060</td>\n",
       "      <td>0.005909</td>\n",
       "      <td>0.043363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>-4.674272e-08</td>\n",
       "      <td>-5.533146e-08</td>\n",
       "      <td>-1.020742e-07</td>\n",
       "      <td>0.041060</td>\n",
       "      <td>0.005909</td>\n",
       "      <td>0.043363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>30</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.041060</td>\n",
       "      <td>0.005909</td>\n",
       "      <td>0.043363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14065</th>\n",
       "      <td>699</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>-8.224614e-10</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-8.224614e-10</td>\n",
       "      <td>-0.007742</td>\n",
       "      <td>0.004884</td>\n",
       "      <td>-0.002695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14066</th>\n",
       "      <td>699</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-3.974721e-04</td>\n",
       "      <td>-2.121521e-07</td>\n",
       "      <td>-3.986902e-04</td>\n",
       "      <td>-0.008139</td>\n",
       "      <td>0.004883</td>\n",
       "      <td>-0.003094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14067</th>\n",
       "      <td>699</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.057307e-07</td>\n",
       "      <td>-1.156271e-07</td>\n",
       "      <td>-2.213577e-07</td>\n",
       "      <td>-0.007742</td>\n",
       "      <td>0.004884</td>\n",
       "      <td>-0.002695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14068</th>\n",
       "      <td>699</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-8.171748e-10</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-8.171748e-10</td>\n",
       "      <td>-0.007742</td>\n",
       "      <td>0.004884</td>\n",
       "      <td>-0.002695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14069</th>\n",
       "      <td>699</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>4.442005e-03</td>\n",
       "      <td>-7.317011e-03</td>\n",
       "      <td>-2.779351e-03</td>\n",
       "      <td>-0.003300</td>\n",
       "      <td>-0.002433</td>\n",
       "      <td>-0.005475</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14070 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator         rep_1         rep_2  \\\n",
       "0        30          A          True -4.105970e-02 -5.908830e-03   \n",
       "1        30          R         False -1.367643e-08 -2.106226e-08   \n",
       "2        30          N         False  0.000000e+00  0.000000e+00   \n",
       "3        30          D         False -4.674272e-08 -5.533146e-08   \n",
       "4        30          C         False  0.000000e+00  0.000000e+00   \n",
       "...     ...        ...           ...           ...           ...   \n",
       "14065   699          T         False -8.224614e-10  0.000000e+00   \n",
       "14066   699          W         False -3.974721e-04 -2.121521e-07   \n",
       "14067   699          Y         False -1.057307e-07 -1.156271e-07   \n",
       "14068   699          V         False -8.171748e-10  0.000000e+00   \n",
       "14069   699          *         False  4.442005e-03 -7.317011e-03   \n",
       "\n",
       "              joint  rep_1_gauged  rep_2_gauged  joint_gauged  \n",
       "0     -4.336264e-02      0.000000      0.000000      0.000000  \n",
       "1     -3.473869e-08      0.041060      0.005909      0.043363  \n",
       "2      0.000000e+00      0.041060      0.005909      0.043363  \n",
       "3     -1.020742e-07      0.041060      0.005909      0.043363  \n",
       "4      0.000000e+00      0.041060      0.005909      0.043363  \n",
       "...             ...           ...           ...           ...  \n",
       "14065 -8.224614e-10     -0.007742      0.004884     -0.002695  \n",
       "14066 -3.986902e-04     -0.008139      0.004883     -0.003094  \n",
       "14067 -2.213577e-07     -0.007742      0.004884     -0.002695  \n",
       "14068 -8.171748e-10     -0.007742      0.004884     -0.002695  \n",
       "14069 -2.779351e-03     -0.003300     -0.002433     -0.005475  \n",
       "\n",
       "[14070 rows x 9 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/rhesus_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/HIV BF520 rhesus host.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/HIV BF520 rhesus host.csv.gz'\n",
    "replicates = 2\n",
    "wildtype(codon_file, selection_file, output_file, replicates)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6aa49d75",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.088512</td>\n",
       "      <td>-0.046735</td>\n",
       "      <td>-0.104449</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.004322</td>\n",
       "      <td>-0.000337</td>\n",
       "      <td>-0.005837</td>\n",
       "      <td>0.084191</td>\n",
       "      <td>0.046398</td>\n",
       "      <td>0.098611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>30</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.004048</td>\n",
       "      <td>-0.004190</td>\n",
       "      <td>-0.008656</td>\n",
       "      <td>0.084465</td>\n",
       "      <td>0.042545</td>\n",
       "      <td>0.095792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>0.002715</td>\n",
       "      <td>0.005175</td>\n",
       "      <td>0.007046</td>\n",
       "      <td>0.091227</td>\n",
       "      <td>0.051910</td>\n",
       "      <td>0.111494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>30</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.003383</td>\n",
       "      <td>0.000357</td>\n",
       "      <td>-0.003288</td>\n",
       "      <td>0.085129</td>\n",
       "      <td>0.047092</td>\n",
       "      <td>0.101161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14065</th>\n",
       "      <td>699</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.006062</td>\n",
       "      <td>0.001052</td>\n",
       "      <td>-0.004876</td>\n",
       "      <td>-0.015892</td>\n",
       "      <td>-0.000213</td>\n",
       "      <td>-0.015090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14066</th>\n",
       "      <td>699</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.004869</td>\n",
       "      <td>-0.002079</td>\n",
       "      <td>-0.006927</td>\n",
       "      <td>-0.014699</td>\n",
       "      <td>-0.003344</td>\n",
       "      <td>-0.017141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14067</th>\n",
       "      <td>699</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.001773</td>\n",
       "      <td>0.006015</td>\n",
       "      <td>0.004309</td>\n",
       "      <td>-0.011604</td>\n",
       "      <td>0.004750</td>\n",
       "      <td>-0.005904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14068</th>\n",
       "      <td>699</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.012105</td>\n",
       "      <td>-0.002908</td>\n",
       "      <td>-0.014847</td>\n",
       "      <td>-0.021935</td>\n",
       "      <td>-0.004174</td>\n",
       "      <td>-0.025060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14069</th>\n",
       "      <td>699</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.006634</td>\n",
       "      <td>-0.006637</td>\n",
       "      <td>-0.013196</td>\n",
       "      <td>-0.016464</td>\n",
       "      <td>-0.007903</td>\n",
       "      <td>-0.023410</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14070 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator     rep_1     rep_2     joint  \\\n",
       "0        30          A          True -0.088512 -0.046735 -0.104449   \n",
       "1        30          R         False -0.004322 -0.000337 -0.005837   \n",
       "2        30          N         False -0.004048 -0.004190 -0.008656   \n",
       "3        30          D         False  0.002715  0.005175  0.007046   \n",
       "4        30          C         False -0.003383  0.000357 -0.003288   \n",
       "...     ...        ...           ...       ...       ...       ...   \n",
       "14065   699          T         False -0.006062  0.001052 -0.004876   \n",
       "14066   699          W         False -0.004869 -0.002079 -0.006927   \n",
       "14067   699          Y         False -0.001773  0.006015  0.004309   \n",
       "14068   699          V         False -0.012105 -0.002908 -0.014847   \n",
       "14069   699          *         False -0.006634 -0.006637 -0.013196   \n",
       "\n",
       "       rep_1_gauged  rep_2_gauged  joint_gauged  \n",
       "0          0.000000      0.000000      0.000000  \n",
       "1          0.084191      0.046398      0.098611  \n",
       "2          0.084465      0.042545      0.095792  \n",
       "3          0.091227      0.051910      0.111494  \n",
       "4          0.085129      0.047092      0.101161  \n",
       "...             ...           ...           ...  \n",
       "14065     -0.015892     -0.000213     -0.015090  \n",
       "14066     -0.014699     -0.003344     -0.017141  \n",
       "14067     -0.011604      0.004750     -0.005904  \n",
       "14068     -0.021935     -0.004174     -0.025060  \n",
       "14069     -0.016464     -0.007903     -0.023410  \n",
       "\n",
       "[14070 rows x 9 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/VRC34_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/HIV bnAbs VRC34.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/HIV bnAbs VRC34.csv.gz'\n",
    "replicates = 2\n",
    "wildtype(codon_file, selection_file, output_file, replicates)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ab95cb93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.080650</td>\n",
       "      <td>-0.029013</td>\n",
       "      <td>-0.083075</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.004672</td>\n",
       "      <td>-0.015205</td>\n",
       "      <td>-0.020216</td>\n",
       "      <td>0.075979</td>\n",
       "      <td>0.013808</td>\n",
       "      <td>0.062859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>30</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.002804</td>\n",
       "      <td>-0.007088</td>\n",
       "      <td>-0.010122</td>\n",
       "      <td>0.077846</td>\n",
       "      <td>0.021925</td>\n",
       "      <td>0.072954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000402</td>\n",
       "      <td>0.001237</td>\n",
       "      <td>0.001124</td>\n",
       "      <td>0.081052</td>\n",
       "      <td>0.030250</td>\n",
       "      <td>0.084200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>30</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.004945</td>\n",
       "      <td>-0.002690</td>\n",
       "      <td>-0.007698</td>\n",
       "      <td>0.075705</td>\n",
       "      <td>0.026323</td>\n",
       "      <td>0.075378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14065</th>\n",
       "      <td>699</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.002445</td>\n",
       "      <td>0.004679</td>\n",
       "      <td>0.002244</td>\n",
       "      <td>-0.000913</td>\n",
       "      <td>-0.000063</td>\n",
       "      <td>-0.000979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14066</th>\n",
       "      <td>699</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.001201</td>\n",
       "      <td>0.000165</td>\n",
       "      <td>-0.001017</td>\n",
       "      <td>0.000331</td>\n",
       "      <td>-0.004577</td>\n",
       "      <td>-0.004241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14067</th>\n",
       "      <td>699</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.001784</td>\n",
       "      <td>0.004086</td>\n",
       "      <td>0.002305</td>\n",
       "      <td>-0.000251</td>\n",
       "      <td>-0.000655</td>\n",
       "      <td>-0.000919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14068</th>\n",
       "      <td>699</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>0.006809</td>\n",
       "      <td>-0.006085</td>\n",
       "      <td>0.000860</td>\n",
       "      <td>0.008341</td>\n",
       "      <td>-0.010827</td>\n",
       "      <td>-0.002364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14069</th>\n",
       "      <td>699</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.006671</td>\n",
       "      <td>-0.006626</td>\n",
       "      <td>-0.013242</td>\n",
       "      <td>-0.005139</td>\n",
       "      <td>-0.011368</td>\n",
       "      <td>-0.016466</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14070 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator     rep_1     rep_2     joint  \\\n",
       "0        30          A          True -0.080650 -0.029013 -0.083075   \n",
       "1        30          R         False -0.004672 -0.015205 -0.020216   \n",
       "2        30          N         False -0.002804 -0.007088 -0.010122   \n",
       "3        30          D         False  0.000402  0.001237  0.001124   \n",
       "4        30          C         False -0.004945 -0.002690 -0.007698   \n",
       "...     ...        ...           ...       ...       ...       ...   \n",
       "14065   699          T         False -0.002445  0.004679  0.002244   \n",
       "14066   699          W         False -0.001201  0.000165 -0.001017   \n",
       "14067   699          Y         False -0.001784  0.004086  0.002305   \n",
       "14068   699          V         False  0.006809 -0.006085  0.000860   \n",
       "14069   699          *         False -0.006671 -0.006626 -0.013242   \n",
       "\n",
       "       rep_1_gauged  rep_2_gauged  joint_gauged  \n",
       "0          0.000000      0.000000      0.000000  \n",
       "1          0.075979      0.013808      0.062859  \n",
       "2          0.077846      0.021925      0.072954  \n",
       "3          0.081052      0.030250      0.084200  \n",
       "4          0.075705      0.026323      0.075378  \n",
       "...             ...           ...           ...  \n",
       "14065     -0.000913     -0.000063     -0.000979  \n",
       "14066      0.000331     -0.004577     -0.004241  \n",
       "14067     -0.000251     -0.000655     -0.000919  \n",
       "14068      0.008341     -0.010827     -0.002364  \n",
       "14069     -0.005139     -0.011368     -0.016466  \n",
       "\n",
       "[14070 rows x 9 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/FP16_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/HIV bnAbs FP16.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/HIV bnAbs FP16.csv.gz'\n",
    "replicates = 2\n",
    "wildtype(codon_file, selection_file, output_file, replicates)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "967fff2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.038100</td>\n",
       "      <td>0.003191</td>\n",
       "      <td>-0.022484</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.001999</td>\n",
       "      <td>-0.009691</td>\n",
       "      <td>-0.011616</td>\n",
       "      <td>0.036101</td>\n",
       "      <td>-0.012882</td>\n",
       "      <td>0.010868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>30</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.005244</td>\n",
       "      <td>-0.001791</td>\n",
       "      <td>-0.007104</td>\n",
       "      <td>0.032856</td>\n",
       "      <td>-0.004982</td>\n",
       "      <td>0.015380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>0.008634</td>\n",
       "      <td>-0.001572</td>\n",
       "      <td>0.007000</td>\n",
       "      <td>0.046734</td>\n",
       "      <td>-0.004763</td>\n",
       "      <td>0.029485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>30</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.005706</td>\n",
       "      <td>0.007516</td>\n",
       "      <td>0.001726</td>\n",
       "      <td>0.032394</td>\n",
       "      <td>0.004326</td>\n",
       "      <td>0.024210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14065</th>\n",
       "      <td>699</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.001084</td>\n",
       "      <td>0.005489</td>\n",
       "      <td>0.004472</td>\n",
       "      <td>-0.021047</td>\n",
       "      <td>0.023266</td>\n",
       "      <td>0.001964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14066</th>\n",
       "      <td>699</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.003872</td>\n",
       "      <td>-0.000047</td>\n",
       "      <td>-0.003920</td>\n",
       "      <td>-0.023834</td>\n",
       "      <td>0.017730</td>\n",
       "      <td>-0.006428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14067</th>\n",
       "      <td>699</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>0.002603</td>\n",
       "      <td>0.003388</td>\n",
       "      <td>0.006024</td>\n",
       "      <td>-0.017359</td>\n",
       "      <td>0.021165</td>\n",
       "      <td>0.003516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14068</th>\n",
       "      <td>699</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.001602</td>\n",
       "      <td>0.005857</td>\n",
       "      <td>0.004215</td>\n",
       "      <td>-0.021564</td>\n",
       "      <td>0.023633</td>\n",
       "      <td>0.001707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14069</th>\n",
       "      <td>699</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.006600</td>\n",
       "      <td>-0.006701</td>\n",
       "      <td>-0.013247</td>\n",
       "      <td>-0.026562</td>\n",
       "      <td>0.011076</td>\n",
       "      <td>-0.015755</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14070 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator     rep_1     rep_2     joint  \\\n",
       "0        30          A          True -0.038100  0.003191 -0.022484   \n",
       "1        30          R         False -0.001999 -0.009691 -0.011616   \n",
       "2        30          N         False -0.005244 -0.001791 -0.007104   \n",
       "3        30          D         False  0.008634 -0.001572  0.007000   \n",
       "4        30          C         False -0.005706  0.007516  0.001726   \n",
       "...     ...        ...           ...       ...       ...       ...   \n",
       "14065   699          T         False -0.001084  0.005489  0.004472   \n",
       "14066   699          W         False -0.003872 -0.000047 -0.003920   \n",
       "14067   699          Y         False  0.002603  0.003388  0.006024   \n",
       "14068   699          V         False -0.001602  0.005857  0.004215   \n",
       "14069   699          *         False -0.006600 -0.006701 -0.013247   \n",
       "\n",
       "       rep_1_gauged  rep_2_gauged  joint_gauged  \n",
       "0          0.000000      0.000000      0.000000  \n",
       "1          0.036101     -0.012882      0.010868  \n",
       "2          0.032856     -0.004982      0.015380  \n",
       "3          0.046734     -0.004763      0.029485  \n",
       "4          0.032394      0.004326      0.024210  \n",
       "...             ...           ...           ...  \n",
       "14065     -0.021047      0.023266      0.001964  \n",
       "14066     -0.023834      0.017730     -0.006428  \n",
       "14067     -0.017359      0.021165      0.003516  \n",
       "14068     -0.021564      0.023633      0.001707  \n",
       "14069     -0.026562      0.011076     -0.015755  \n",
       "\n",
       "[14070 rows x 9 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/FP20_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/HIV bnAbs FP20.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/HIV bnAbs FP20.csv.gz'\n",
    "replicates = 2\n",
    "wildtype(codon_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c5bcbfbf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>rep_3</th>\n",
       "      <th>rep_4</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>rep_3_gauged</th>\n",
       "      <th>rep_4_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.228206</td>\n",
       "      <td>-0.169474</td>\n",
       "      <td>-0.189958</td>\n",
       "      <td>0.087909</td>\n",
       "      <td>-0.222629</td>\n",
       "      <td>-0.388529</td>\n",
       "      <td>-0.659860</td>\n",
       "      <td>-0.603675</td>\n",
       "      <td>-0.141804</td>\n",
       "      <td>-0.563405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>0.121223</td>\n",
       "      <td>-0.041648</td>\n",
       "      <td>0.090308</td>\n",
       "      <td>-0.112834</td>\n",
       "      <td>0.051511</td>\n",
       "      <td>-0.039100</td>\n",
       "      <td>-0.532035</td>\n",
       "      <td>-0.323409</td>\n",
       "      <td>-0.342546</td>\n",
       "      <td>-0.289265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>0.077750</td>\n",
       "      <td>-0.009723</td>\n",
       "      <td>0.234759</td>\n",
       "      <td>0.005293</td>\n",
       "      <td>0.185005</td>\n",
       "      <td>-0.082573</td>\n",
       "      <td>-0.500109</td>\n",
       "      <td>-0.178959</td>\n",
       "      <td>-0.224419</td>\n",
       "      <td>-0.155770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>0.034013</td>\n",
       "      <td>0.304593</td>\n",
       "      <td>-0.032153</td>\n",
       "      <td>0.189254</td>\n",
       "      <td>0.301517</td>\n",
       "      <td>-0.126310</td>\n",
       "      <td>-0.185794</td>\n",
       "      <td>-0.445871</td>\n",
       "      <td>-0.040459</td>\n",
       "      <td>-0.039259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.110994</td>\n",
       "      <td>0.100525</td>\n",
       "      <td>0.058474</td>\n",
       "      <td>0.080342</td>\n",
       "      <td>0.207113</td>\n",
       "      <td>-0.049329</td>\n",
       "      <td>-0.389861</td>\n",
       "      <td>-0.355244</td>\n",
       "      <td>-0.149371</td>\n",
       "      <td>-0.133662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11902</th>\n",
       "      <td>567</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.043499</td>\n",
       "      <td>-0.096243</td>\n",
       "      <td>-0.053968</td>\n",
       "      <td>-0.102841</td>\n",
       "      <td>-0.117557</td>\n",
       "      <td>-0.891262</td>\n",
       "      <td>-1.183119</td>\n",
       "      <td>-1.176356</td>\n",
       "      <td>-1.163957</td>\n",
       "      <td>-1.406100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11903</th>\n",
       "      <td>567</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>0.095643</td>\n",
       "      <td>0.219418</td>\n",
       "      <td>0.039788</td>\n",
       "      <td>0.161092</td>\n",
       "      <td>0.466630</td>\n",
       "      <td>-0.752119</td>\n",
       "      <td>-0.867458</td>\n",
       "      <td>-1.082599</td>\n",
       "      <td>-0.900024</td>\n",
       "      <td>-0.821914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11904</th>\n",
       "      <td>567</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>0.004296</td>\n",
       "      <td>-0.024369</td>\n",
       "      <td>-0.023013</td>\n",
       "      <td>-0.036342</td>\n",
       "      <td>-0.004656</td>\n",
       "      <td>-0.843466</td>\n",
       "      <td>-1.111244</td>\n",
       "      <td>-1.145401</td>\n",
       "      <td>-1.097457</td>\n",
       "      <td>-1.293199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11905</th>\n",
       "      <td>567</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.124261</td>\n",
       "      <td>-0.106221</td>\n",
       "      <td>-0.014119</td>\n",
       "      <td>-0.081844</td>\n",
       "      <td>-0.122425</td>\n",
       "      <td>-0.972023</td>\n",
       "      <td>-1.193096</td>\n",
       "      <td>-1.136507</td>\n",
       "      <td>-1.142959</td>\n",
       "      <td>-1.410968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11906</th>\n",
       "      <td>567</td>\n",
       "      <td>*</td>\n",
       "      <td>True</td>\n",
       "      <td>0.847763</td>\n",
       "      <td>1.086875</td>\n",
       "      <td>1.122388</td>\n",
       "      <td>1.061116</td>\n",
       "      <td>1.288543</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11907 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator     rep_1     rep_2     rep_3     rep_4  \\\n",
       "0         1          A         False -0.228206 -0.169474 -0.189958  0.087909   \n",
       "1         1          R         False  0.121223 -0.041648  0.090308 -0.112834   \n",
       "2         1          N         False  0.077750 -0.009723  0.234759  0.005293   \n",
       "3         1          D         False  0.034013  0.304593 -0.032153  0.189254   \n",
       "4         1          C         False  0.110994  0.100525  0.058474  0.080342   \n",
       "...     ...        ...           ...       ...       ...       ...       ...   \n",
       "11902   567          T         False -0.043499 -0.096243 -0.053968 -0.102841   \n",
       "11903   567          W         False  0.095643  0.219418  0.039788  0.161092   \n",
       "11904   567          Y         False  0.004296 -0.024369 -0.023013 -0.036342   \n",
       "11905   567          V         False -0.124261 -0.106221 -0.014119 -0.081844   \n",
       "11906   567          *          True  0.847763  1.086875  1.122388  1.061116   \n",
       "\n",
       "          joint  rep_1_gauged  rep_2_gauged  rep_3_gauged  rep_4_gauged  \\\n",
       "0     -0.222629     -0.388529     -0.659860     -0.603675     -0.141804   \n",
       "1      0.051511     -0.039100     -0.532035     -0.323409     -0.342546   \n",
       "2      0.185005     -0.082573     -0.500109     -0.178959     -0.224419   \n",
       "3      0.301517     -0.126310     -0.185794     -0.445871     -0.040459   \n",
       "4      0.207113     -0.049329     -0.389861     -0.355244     -0.149371   \n",
       "...         ...           ...           ...           ...           ...   \n",
       "11902 -0.117557     -0.891262     -1.183119     -1.176356     -1.163957   \n",
       "11903  0.466630     -0.752119     -0.867458     -1.082599     -0.900024   \n",
       "11904 -0.004656     -0.843466     -1.111244     -1.145401     -1.097457   \n",
       "11905 -0.122425     -0.972023     -1.193096     -1.136507     -1.142959   \n",
       "11906  1.288543      0.000000      0.000000      0.000000      0.000000   \n",
       "\n",
       "       joint_gauged  \n",
       "0         -0.563405  \n",
       "1         -0.289265  \n",
       "2         -0.155770  \n",
       "3         -0.039259  \n",
       "4         -0.133662  \n",
       "...             ...  \n",
       "11902     -1.406100  \n",
       "11903     -0.821914  \n",
       "11904     -1.293199  \n",
       "11905     -1.410968  \n",
       "11906      0.000000  \n",
       "\n",
       "[11907 rows x 13 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/Perth2009_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/Perth2009.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/Perth2009.csv.gz'\n",
    "replicates = 4\n",
    "wildtype(codon_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9084ff76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>rep_3</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>rep_3_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.061280</td>\n",
       "      <td>-0.049304</td>\n",
       "      <td>-0.047489</td>\n",
       "      <td>-0.088229</td>\n",
       "      <td>-0.772519</td>\n",
       "      <td>-0.711617</td>\n",
       "      <td>-0.725481</td>\n",
       "      <td>-1.150491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.089901</td>\n",
       "      <td>-0.073764</td>\n",
       "      <td>-0.085586</td>\n",
       "      <td>-0.159624</td>\n",
       "      <td>-0.801139</td>\n",
       "      <td>-0.736077</td>\n",
       "      <td>-0.763578</td>\n",
       "      <td>-1.221886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.037012</td>\n",
       "      <td>-0.031161</td>\n",
       "      <td>-0.020695</td>\n",
       "      <td>-0.058393</td>\n",
       "      <td>-0.748251</td>\n",
       "      <td>-0.693474</td>\n",
       "      <td>-0.698687</td>\n",
       "      <td>-1.120655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.025307</td>\n",
       "      <td>-0.019428</td>\n",
       "      <td>-0.023522</td>\n",
       "      <td>-0.047132</td>\n",
       "      <td>-0.736545</td>\n",
       "      <td>-0.681741</td>\n",
       "      <td>-0.701514</td>\n",
       "      <td>-1.109394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.026486</td>\n",
       "      <td>-0.018230</td>\n",
       "      <td>-0.033550</td>\n",
       "      <td>-0.053815</td>\n",
       "      <td>-0.737725</td>\n",
       "      <td>-0.680543</td>\n",
       "      <td>-0.711542</td>\n",
       "      <td>-1.116077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15955</th>\n",
       "      <td>760</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.059244</td>\n",
       "      <td>-0.059847</td>\n",
       "      <td>-0.076727</td>\n",
       "      <td>-0.120981</td>\n",
       "      <td>-0.895433</td>\n",
       "      <td>-0.815671</td>\n",
       "      <td>-0.852086</td>\n",
       "      <td>-1.293460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15956</th>\n",
       "      <td>760</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>0.026390</td>\n",
       "      <td>0.041105</td>\n",
       "      <td>0.056984</td>\n",
       "      <td>0.178988</td>\n",
       "      <td>-0.809799</td>\n",
       "      <td>-0.714718</td>\n",
       "      <td>-0.718375</td>\n",
       "      <td>-0.993491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15957</th>\n",
       "      <td>760</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.055756</td>\n",
       "      <td>-0.042967</td>\n",
       "      <td>-0.042822</td>\n",
       "      <td>-0.079030</td>\n",
       "      <td>-0.891946</td>\n",
       "      <td>-0.798790</td>\n",
       "      <td>-0.818181</td>\n",
       "      <td>-1.251509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15958</th>\n",
       "      <td>760</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.071218</td>\n",
       "      <td>-0.066983</td>\n",
       "      <td>-0.057044</td>\n",
       "      <td>-0.115528</td>\n",
       "      <td>-0.907408</td>\n",
       "      <td>-0.822807</td>\n",
       "      <td>-0.832403</td>\n",
       "      <td>-1.288006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15959</th>\n",
       "      <td>760</td>\n",
       "      <td>*</td>\n",
       "      <td>True</td>\n",
       "      <td>0.836190</td>\n",
       "      <td>0.755824</td>\n",
       "      <td>0.775359</td>\n",
       "      <td>1.172479</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15960 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator     rep_1     rep_2     rep_3     joint  \\\n",
       "0         1          A         False -0.061280 -0.049304 -0.047489 -0.088229   \n",
       "1         1          R         False -0.089901 -0.073764 -0.085586 -0.159624   \n",
       "2         1          N         False -0.037012 -0.031161 -0.020695 -0.058393   \n",
       "3         1          D         False -0.025307 -0.019428 -0.023522 -0.047132   \n",
       "4         1          C         False -0.026486 -0.018230 -0.033550 -0.053815   \n",
       "...     ...        ...           ...       ...       ...       ...       ...   \n",
       "15955   760          T         False -0.059244 -0.059847 -0.076727 -0.120981   \n",
       "15956   760          W         False  0.026390  0.041105  0.056984  0.178988   \n",
       "15957   760          Y         False -0.055756 -0.042967 -0.042822 -0.079030   \n",
       "15958   760          V         False -0.071218 -0.066983 -0.057044 -0.115528   \n",
       "15959   760          *          True  0.836190  0.755824  0.775359  1.172479   \n",
       "\n",
       "       rep_1_gauged  rep_2_gauged  rep_3_gauged  joint_gauged  \n",
       "0         -0.772519     -0.711617     -0.725481     -1.150491  \n",
       "1         -0.801139     -0.736077     -0.763578     -1.221886  \n",
       "2         -0.748251     -0.693474     -0.698687     -1.120655  \n",
       "3         -0.736545     -0.681741     -0.701514     -1.109394  \n",
       "4         -0.737725     -0.680543     -0.711542     -1.116077  \n",
       "...             ...           ...           ...           ...  \n",
       "15955     -0.895433     -0.815671     -0.852086     -1.293460  \n",
       "15956     -0.809799     -0.714718     -0.718375     -0.993491  \n",
       "15957     -0.891946     -0.798790     -0.818181     -1.251509  \n",
       "15958     -0.907408     -0.822807     -0.832403     -1.288006  \n",
       "15959      0.000000      0.000000      0.000000      0.000000  \n",
       "\n",
       "[15960 rows x 11 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/CCL141_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/CCL141.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/CCL141.csv.gz'\n",
    "replicates = 3\n",
    "wildtype(codon_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6d096bae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>rep_3</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>rep_3_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.069898</td>\n",
       "      <td>-0.048427</td>\n",
       "      <td>-0.058717</td>\n",
       "      <td>-0.103474</td>\n",
       "      <td>-0.809507</td>\n",
       "      <td>-0.775116</td>\n",
       "      <td>-0.797766</td>\n",
       "      <td>-1.255812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.085384</td>\n",
       "      <td>-0.069942</td>\n",
       "      <td>-0.076981</td>\n",
       "      <td>-0.136801</td>\n",
       "      <td>-0.824993</td>\n",
       "      <td>-0.796632</td>\n",
       "      <td>-0.816030</td>\n",
       "      <td>-1.289138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.036199</td>\n",
       "      <td>-0.032016</td>\n",
       "      <td>-0.022438</td>\n",
       "      <td>-0.058073</td>\n",
       "      <td>-0.775808</td>\n",
       "      <td>-0.758706</td>\n",
       "      <td>-0.761487</td>\n",
       "      <td>-1.210411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.019406</td>\n",
       "      <td>-0.018494</td>\n",
       "      <td>-0.022437</td>\n",
       "      <td>-0.036630</td>\n",
       "      <td>-0.759015</td>\n",
       "      <td>-0.745184</td>\n",
       "      <td>-0.761486</td>\n",
       "      <td>-1.188968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.025904</td>\n",
       "      <td>-0.017354</td>\n",
       "      <td>-0.032003</td>\n",
       "      <td>-0.048657</td>\n",
       "      <td>-0.765513</td>\n",
       "      <td>-0.744044</td>\n",
       "      <td>-0.771052</td>\n",
       "      <td>-1.200994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15955</th>\n",
       "      <td>760</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.058473</td>\n",
       "      <td>-0.061532</td>\n",
       "      <td>-0.075577</td>\n",
       "      <td>-0.121429</td>\n",
       "      <td>-0.909837</td>\n",
       "      <td>-0.782275</td>\n",
       "      <td>-0.869323</td>\n",
       "      <td>-1.290837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15956</th>\n",
       "      <td>760</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>0.039764</td>\n",
       "      <td>0.032528</td>\n",
       "      <td>0.044873</td>\n",
       "      <td>0.171268</td>\n",
       "      <td>-0.811600</td>\n",
       "      <td>-0.688215</td>\n",
       "      <td>-0.748873</td>\n",
       "      <td>-0.998140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15957</th>\n",
       "      <td>760</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.056840</td>\n",
       "      <td>-0.057206</td>\n",
       "      <td>-0.034837</td>\n",
       "      <td>-0.087070</td>\n",
       "      <td>-0.908204</td>\n",
       "      <td>-0.777949</td>\n",
       "      <td>-0.828583</td>\n",
       "      <td>-1.256478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15958</th>\n",
       "      <td>760</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.065433</td>\n",
       "      <td>-0.058801</td>\n",
       "      <td>-0.054878</td>\n",
       "      <td>-0.099144</td>\n",
       "      <td>-0.916797</td>\n",
       "      <td>-0.779544</td>\n",
       "      <td>-0.848623</td>\n",
       "      <td>-1.268552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15959</th>\n",
       "      <td>760</td>\n",
       "      <td>*</td>\n",
       "      <td>True</td>\n",
       "      <td>0.851364</td>\n",
       "      <td>0.720743</td>\n",
       "      <td>0.793746</td>\n",
       "      <td>1.169408</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15960 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator     rep_1     rep_2     rep_3     joint  \\\n",
       "0         1          A         False -0.069898 -0.048427 -0.058717 -0.103474   \n",
       "1         1          R         False -0.085384 -0.069942 -0.076981 -0.136801   \n",
       "2         1          N         False -0.036199 -0.032016 -0.022438 -0.058073   \n",
       "3         1          D         False -0.019406 -0.018494 -0.022437 -0.036630   \n",
       "4         1          C         False -0.025904 -0.017354 -0.032003 -0.048657   \n",
       "...     ...        ...           ...       ...       ...       ...       ...   \n",
       "15955   760          T         False -0.058473 -0.061532 -0.075577 -0.121429   \n",
       "15956   760          W         False  0.039764  0.032528  0.044873  0.171268   \n",
       "15957   760          Y         False -0.056840 -0.057206 -0.034837 -0.087070   \n",
       "15958   760          V         False -0.065433 -0.058801 -0.054878 -0.099144   \n",
       "15959   760          *          True  0.851364  0.720743  0.793746  1.169408   \n",
       "\n",
       "       rep_1_gauged  rep_2_gauged  rep_3_gauged  joint_gauged  \n",
       "0         -0.809507     -0.775116     -0.797766     -1.255812  \n",
       "1         -0.824993     -0.796632     -0.816030     -1.289138  \n",
       "2         -0.775808     -0.758706     -0.761487     -1.210411  \n",
       "3         -0.759015     -0.745184     -0.761486     -1.188968  \n",
       "4         -0.765513     -0.744044     -0.771052     -1.200994  \n",
       "...             ...           ...           ...           ...  \n",
       "15955     -0.909837     -0.782275     -0.869323     -1.290837  \n",
       "15956     -0.811600     -0.688215     -0.748873     -0.998140  \n",
       "15957     -0.908204     -0.777949     -0.828583     -1.256478  \n",
       "15958     -0.916797     -0.779544     -0.848623     -1.268552  \n",
       "15959      0.000000      0.000000      0.000000      0.000000  \n",
       "\n",
       "[15960 rows x 11 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/A549_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/A549.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/A549.csv.gz'\n",
    "replicates = 3\n",
    "wildtype(codon_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c1cde8e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.642490e-06</td>\n",
       "      <td>-1.022034e-06</td>\n",
       "      <td>-2.664524e-06</td>\n",
       "      <td>-0.280304</td>\n",
       "      <td>-0.265423</td>\n",
       "      <td>-0.429581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>3.771611e-03</td>\n",
       "      <td>9.100634e-03</td>\n",
       "      <td>1.451313e-02</td>\n",
       "      <td>-0.276531</td>\n",
       "      <td>-0.256321</td>\n",
       "      <td>-0.415065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-6.355282e-07</td>\n",
       "      <td>-4.109946e-07</td>\n",
       "      <td>-1.046523e-06</td>\n",
       "      <td>-0.280303</td>\n",
       "      <td>-0.265422</td>\n",
       "      <td>-0.429579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-5.827963e-09</td>\n",
       "      <td>-5.827963e-09</td>\n",
       "      <td>-0.280302</td>\n",
       "      <td>-0.265422</td>\n",
       "      <td>-0.429578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>-3.727336e-09</td>\n",
       "      <td>-3.885309e-09</td>\n",
       "      <td>-7.612645e-09</td>\n",
       "      <td>-0.280302</td>\n",
       "      <td>-0.265422</td>\n",
       "      <td>-0.429578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10453</th>\n",
       "      <td>498</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>3.979515e-01</td>\n",
       "      <td>3.544422e-01</td>\n",
       "      <td>5.340280e-01</td>\n",
       "      <td>-0.455157</td>\n",
       "      <td>-0.206273</td>\n",
       "      <td>-0.249650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10454</th>\n",
       "      <td>498</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-3.222397e-01</td>\n",
       "      <td>-3.782226e-01</td>\n",
       "      <td>-5.122538e-01</td>\n",
       "      <td>-1.175348</td>\n",
       "      <td>-0.938938</td>\n",
       "      <td>-1.295931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10455</th>\n",
       "      <td>498</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-2.442741e-01</td>\n",
       "      <td>-2.977706e-01</td>\n",
       "      <td>-4.134223e-01</td>\n",
       "      <td>-1.097383</td>\n",
       "      <td>-0.858486</td>\n",
       "      <td>-1.197100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10456</th>\n",
       "      <td>498</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-5.250783e-01</td>\n",
       "      <td>-5.549608e-01</td>\n",
       "      <td>-6.357450e-01</td>\n",
       "      <td>-1.378187</td>\n",
       "      <td>-1.115676</td>\n",
       "      <td>-1.419423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10457</th>\n",
       "      <td>498</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>4.123651e-01</td>\n",
       "      <td>-3.838369e-01</td>\n",
       "      <td>1.117703e-01</td>\n",
       "      <td>-0.440743</td>\n",
       "      <td>-0.944552</td>\n",
       "      <td>-0.671907</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10458 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator         rep_1         rep_2  \\\n",
       "0         1          A         False -1.642490e-06 -1.022034e-06   \n",
       "1         1          R         False  3.771611e-03  9.100634e-03   \n",
       "2         1          N         False -6.355282e-07 -4.109946e-07   \n",
       "3         1          D         False  0.000000e+00 -5.827963e-09   \n",
       "4         1          C         False -3.727336e-09 -3.885309e-09   \n",
       "...     ...        ...           ...           ...           ...   \n",
       "10453   498          T         False  3.979515e-01  3.544422e-01   \n",
       "10454   498          W         False -3.222397e-01 -3.782226e-01   \n",
       "10455   498          Y         False -2.442741e-01 -2.977706e-01   \n",
       "10456   498          V         False -5.250783e-01 -5.549608e-01   \n",
       "10457   498          *         False  4.123651e-01 -3.838369e-01   \n",
       "\n",
       "              joint  rep_1_gauged  rep_2_gauged  joint_gauged  \n",
       "0     -2.664524e-06     -0.280304     -0.265423     -0.429581  \n",
       "1      1.451313e-02     -0.276531     -0.256321     -0.415065  \n",
       "2     -1.046523e-06     -0.280303     -0.265422     -0.429579  \n",
       "3     -5.827963e-09     -0.280302     -0.265422     -0.429578  \n",
       "4     -7.612645e-09     -0.280302     -0.265422     -0.429578  \n",
       "...             ...           ...           ...           ...  \n",
       "10453  5.340280e-01     -0.455157     -0.206273     -0.249650  \n",
       "10454 -5.122538e-01     -1.175348     -0.938938     -1.295931  \n",
       "10455 -4.134223e-01     -1.097383     -0.858486     -1.197100  \n",
       "10456 -6.357450e-01     -1.378187     -1.115676     -1.419423  \n",
       "10457  1.117703e-01     -0.440743     -0.944552     -0.671907  \n",
       "\n",
       "[10458 rows x 9 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/MxA_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/MxA.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/MxA.csv.gz'\n",
    "replicates = 2\n",
    "wildtype(codon_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f594055a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.642490e-06</td>\n",
       "      <td>-1.022034e-06</td>\n",
       "      <td>-2.664524e-06</td>\n",
       "      <td>-0.237307</td>\n",
       "      <td>-0.233431</td>\n",
       "      <td>-0.364074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>7.254972e-03</td>\n",
       "      <td>1.002249e-02</td>\n",
       "      <td>1.881476e-02</td>\n",
       "      <td>-0.230051</td>\n",
       "      <td>-0.223408</td>\n",
       "      <td>-0.345257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-6.355282e-07</td>\n",
       "      <td>-4.109946e-07</td>\n",
       "      <td>-1.046523e-06</td>\n",
       "      <td>-0.237306</td>\n",
       "      <td>-0.233431</td>\n",
       "      <td>-0.364073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-5.827963e-09</td>\n",
       "      <td>-5.827963e-09</td>\n",
       "      <td>-0.237306</td>\n",
       "      <td>-0.233430</td>\n",
       "      <td>-0.364072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>2.902274e-03</td>\n",
       "      <td>-3.885309e-09</td>\n",
       "      <td>3.066658e-03</td>\n",
       "      <td>-0.234403</td>\n",
       "      <td>-0.233430</td>\n",
       "      <td>-0.361005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10453</th>\n",
       "      <td>498</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>5.013515e-01</td>\n",
       "      <td>3.558763e-01</td>\n",
       "      <td>5.944799e-01</td>\n",
       "      <td>-0.367383</td>\n",
       "      <td>-0.222222</td>\n",
       "      <td>-0.201849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10454</th>\n",
       "      <td>498</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-3.031567e-01</td>\n",
       "      <td>-3.754200e-01</td>\n",
       "      <td>-4.964028e-01</td>\n",
       "      <td>-1.171892</td>\n",
       "      <td>-0.953519</td>\n",
       "      <td>-1.292732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10455</th>\n",
       "      <td>498</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-2.576910e-01</td>\n",
       "      <td>-2.718026e-01</td>\n",
       "      <td>-4.031868e-01</td>\n",
       "      <td>-1.126426</td>\n",
       "      <td>-0.849901</td>\n",
       "      <td>-1.199516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10456</th>\n",
       "      <td>498</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-4.746157e-01</td>\n",
       "      <td>-4.938968e-01</td>\n",
       "      <td>-5.653539e-01</td>\n",
       "      <td>-1.343351</td>\n",
       "      <td>-1.071995</td>\n",
       "      <td>-1.361683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10457</th>\n",
       "      <td>498</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>1.326844e-01</td>\n",
       "      <td>-4.280665e-01</td>\n",
       "      <td>-1.553986e-01</td>\n",
       "      <td>-0.736050</td>\n",
       "      <td>-1.006165</td>\n",
       "      <td>-0.951727</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10458 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator         rep_1         rep_2  \\\n",
       "0         1          A         False -1.642490e-06 -1.022034e-06   \n",
       "1         1          R         False  7.254972e-03  1.002249e-02   \n",
       "2         1          N         False -6.355282e-07 -4.109946e-07   \n",
       "3         1          D         False  0.000000e+00 -5.827963e-09   \n",
       "4         1          C         False  2.902274e-03 -3.885309e-09   \n",
       "...     ...        ...           ...           ...           ...   \n",
       "10453   498          T         False  5.013515e-01  3.558763e-01   \n",
       "10454   498          W         False -3.031567e-01 -3.754200e-01   \n",
       "10455   498          Y         False -2.576910e-01 -2.718026e-01   \n",
       "10456   498          V         False -4.746157e-01 -4.938968e-01   \n",
       "10457   498          *         False  1.326844e-01 -4.280665e-01   \n",
       "\n",
       "              joint  rep_1_gauged  rep_2_gauged  joint_gauged  \n",
       "0     -2.664524e-06     -0.237307     -0.233431     -0.364074  \n",
       "1      1.881476e-02     -0.230051     -0.223408     -0.345257  \n",
       "2     -1.046523e-06     -0.237306     -0.233431     -0.364073  \n",
       "3     -5.827963e-09     -0.237306     -0.233430     -0.364072  \n",
       "4      3.066658e-03     -0.234403     -0.233430     -0.361005  \n",
       "...             ...           ...           ...           ...  \n",
       "10453  5.944799e-01     -0.367383     -0.222222     -0.201849  \n",
       "10454 -4.964028e-01     -1.171892     -0.953519     -1.292732  \n",
       "10455 -4.031868e-01     -1.126426     -0.849901     -1.199516  \n",
       "10456 -5.653539e-01     -1.343351     -1.071995     -1.361683  \n",
       "10457 -1.553986e-01     -0.736050     -1.006165     -0.951727  \n",
       "\n",
       "[10458 rows x 9 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/MS_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/MS.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/MS.csv.gz'\n",
    "replicates = 2\n",
    "wildtype(codon_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4dd7bc8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.642490e-06</td>\n",
       "      <td>-1.022034e-06</td>\n",
       "      <td>-2.664524e-06</td>\n",
       "      <td>-0.280304</td>\n",
       "      <td>-0.265423</td>\n",
       "      <td>-0.429581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>3.771611e-03</td>\n",
       "      <td>9.100634e-03</td>\n",
       "      <td>1.451313e-02</td>\n",
       "      <td>-0.276531</td>\n",
       "      <td>-0.256321</td>\n",
       "      <td>-0.415065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-6.355282e-07</td>\n",
       "      <td>-4.109946e-07</td>\n",
       "      <td>-1.046523e-06</td>\n",
       "      <td>-0.280303</td>\n",
       "      <td>-0.265422</td>\n",
       "      <td>-0.429579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-5.827963e-09</td>\n",
       "      <td>-5.827963e-09</td>\n",
       "      <td>-0.280302</td>\n",
       "      <td>-0.265422</td>\n",
       "      <td>-0.429578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>-3.727336e-09</td>\n",
       "      <td>-3.885309e-09</td>\n",
       "      <td>-7.612645e-09</td>\n",
       "      <td>-0.280302</td>\n",
       "      <td>-0.265422</td>\n",
       "      <td>-0.429578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10453</th>\n",
       "      <td>498</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>3.979515e-01</td>\n",
       "      <td>3.544422e-01</td>\n",
       "      <td>5.340280e-01</td>\n",
       "      <td>-0.455157</td>\n",
       "      <td>-0.206273</td>\n",
       "      <td>-0.249650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10454</th>\n",
       "      <td>498</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-3.222397e-01</td>\n",
       "      <td>-3.782226e-01</td>\n",
       "      <td>-5.122538e-01</td>\n",
       "      <td>-1.175348</td>\n",
       "      <td>-0.938938</td>\n",
       "      <td>-1.295931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10455</th>\n",
       "      <td>498</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-2.442741e-01</td>\n",
       "      <td>-2.977706e-01</td>\n",
       "      <td>-4.134223e-01</td>\n",
       "      <td>-1.097383</td>\n",
       "      <td>-0.858486</td>\n",
       "      <td>-1.197100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10456</th>\n",
       "      <td>498</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-5.250783e-01</td>\n",
       "      <td>-5.549608e-01</td>\n",
       "      <td>-6.357450e-01</td>\n",
       "      <td>-1.378187</td>\n",
       "      <td>-1.115676</td>\n",
       "      <td>-1.419423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10457</th>\n",
       "      <td>498</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>4.123651e-01</td>\n",
       "      <td>-3.838369e-01</td>\n",
       "      <td>1.117703e-01</td>\n",
       "      <td>-0.440743</td>\n",
       "      <td>-0.944552</td>\n",
       "      <td>-0.671907</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10458 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator         rep_1         rep_2  \\\n",
       "0         1          A         False -1.642490e-06 -1.022034e-06   \n",
       "1         1          R         False  3.771611e-03  9.100634e-03   \n",
       "2         1          N         False -6.355282e-07 -4.109946e-07   \n",
       "3         1          D         False  0.000000e+00 -5.827963e-09   \n",
       "4         1          C         False -3.727336e-09 -3.885309e-09   \n",
       "...     ...        ...           ...           ...           ...   \n",
       "10453   498          T         False  3.979515e-01  3.544422e-01   \n",
       "10454   498          W         False -3.222397e-01 -3.782226e-01   \n",
       "10455   498          Y         False -2.442741e-01 -2.977706e-01   \n",
       "10456   498          V         False -5.250783e-01 -5.549608e-01   \n",
       "10457   498          *         False  4.123651e-01 -3.838369e-01   \n",
       "\n",
       "              joint  rep_1_gauged  rep_2_gauged  joint_gauged  \n",
       "0     -2.664524e-06     -0.280304     -0.265423     -0.429581  \n",
       "1      1.451313e-02     -0.276531     -0.256321     -0.415065  \n",
       "2     -1.046523e-06     -0.280303     -0.265422     -0.429579  \n",
       "3     -5.827963e-09     -0.280302     -0.265422     -0.429578  \n",
       "4     -7.612645e-09     -0.280302     -0.265422     -0.429578  \n",
       "...             ...           ...           ...           ...  \n",
       "10453  5.340280e-01     -0.455157     -0.206273     -0.249650  \n",
       "10454 -5.122538e-01     -1.175348     -0.938938     -1.295931  \n",
       "10455 -4.134223e-01     -1.097383     -0.858486     -1.197100  \n",
       "10456 -6.357450e-01     -1.378187     -1.115676     -1.419423  \n",
       "10457  1.117703e-01     -0.440743     -0.944552     -0.671907  \n",
       "\n",
       "[10458 rows x 9 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/MxAneg_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/MxAneg.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/MxAneg.csv.gz'\n",
    "replicates = 2\n",
    "wildtype(codon_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4107dc01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>rep_3</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>rep_3_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.143420</td>\n",
       "      <td>-0.130483</td>\n",
       "      <td>-0.155258</td>\n",
       "      <td>-0.213219</td>\n",
       "      <td>-1.397071</td>\n",
       "      <td>-1.381862</td>\n",
       "      <td>-1.378743</td>\n",
       "      <td>-1.708042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.157189</td>\n",
       "      <td>-0.146008</td>\n",
       "      <td>-0.153729</td>\n",
       "      <td>-0.215293</td>\n",
       "      <td>-1.410840</td>\n",
       "      <td>-1.397387</td>\n",
       "      <td>-1.377214</td>\n",
       "      <td>-1.710116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.056040</td>\n",
       "      <td>-0.060483</td>\n",
       "      <td>-0.057589</td>\n",
       "      <td>-0.097759</td>\n",
       "      <td>-1.309691</td>\n",
       "      <td>-1.311862</td>\n",
       "      <td>-1.281074</td>\n",
       "      <td>-1.592582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.061209</td>\n",
       "      <td>-0.059326</td>\n",
       "      <td>-0.051964</td>\n",
       "      <td>-0.096627</td>\n",
       "      <td>-1.314860</td>\n",
       "      <td>-1.310705</td>\n",
       "      <td>-1.275450</td>\n",
       "      <td>-1.591451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.052824</td>\n",
       "      <td>-0.048705</td>\n",
       "      <td>-0.047273</td>\n",
       "      <td>-0.080897</td>\n",
       "      <td>-1.306474</td>\n",
       "      <td>-1.300084</td>\n",
       "      <td>-1.270758</td>\n",
       "      <td>-1.575720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5308</th>\n",
       "      <td>253</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.057007</td>\n",
       "      <td>-0.064913</td>\n",
       "      <td>-0.044685</td>\n",
       "      <td>-0.099932</td>\n",
       "      <td>-0.774206</td>\n",
       "      <td>-0.822459</td>\n",
       "      <td>-0.711495</td>\n",
       "      <td>-1.229170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5309</th>\n",
       "      <td>253</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.003961</td>\n",
       "      <td>-0.009725</td>\n",
       "      <td>0.002024</td>\n",
       "      <td>0.007765</td>\n",
       "      <td>-0.721160</td>\n",
       "      <td>-0.767271</td>\n",
       "      <td>-0.664786</td>\n",
       "      <td>-1.121473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5310</th>\n",
       "      <td>253</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.009920</td>\n",
       "      <td>-0.016659</td>\n",
       "      <td>-0.026625</td>\n",
       "      <td>-0.021214</td>\n",
       "      <td>-0.727119</td>\n",
       "      <td>-0.774205</td>\n",
       "      <td>-0.693435</td>\n",
       "      <td>-1.150451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5311</th>\n",
       "      <td>253</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.043550</td>\n",
       "      <td>-0.038977</td>\n",
       "      <td>-0.045208</td>\n",
       "      <td>-0.078253</td>\n",
       "      <td>-0.760749</td>\n",
       "      <td>-0.796523</td>\n",
       "      <td>-0.712018</td>\n",
       "      <td>-1.207490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5312</th>\n",
       "      <td>253</td>\n",
       "      <td>*</td>\n",
       "      <td>True</td>\n",
       "      <td>0.717199</td>\n",
       "      <td>0.757546</td>\n",
       "      <td>0.666810</td>\n",
       "      <td>1.129237</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5313 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      site amino_acid  WT_indicator     rep_1     rep_2     rep_3     joint  \\\n",
       "0        1          A         False -0.143420 -0.130483 -0.155258 -0.213219   \n",
       "1        1          R         False -0.157189 -0.146008 -0.153729 -0.215293   \n",
       "2        1          N         False -0.056040 -0.060483 -0.057589 -0.097759   \n",
       "3        1          D         False -0.061209 -0.059326 -0.051964 -0.096627   \n",
       "4        1          C         False -0.052824 -0.048705 -0.047273 -0.080897   \n",
       "...    ...        ...           ...       ...       ...       ...       ...   \n",
       "5308   253          T         False -0.057007 -0.064913 -0.044685 -0.099932   \n",
       "5309   253          W         False -0.003961 -0.009725  0.002024  0.007765   \n",
       "5310   253          Y         False -0.009920 -0.016659 -0.026625 -0.021214   \n",
       "5311   253          V         False -0.043550 -0.038977 -0.045208 -0.078253   \n",
       "5312   253          *          True  0.717199  0.757546  0.666810  1.129237   \n",
       "\n",
       "      rep_1_gauged  rep_2_gauged  rep_3_gauged  joint_gauged  \n",
       "0        -1.397071     -1.381862     -1.378743     -1.708042  \n",
       "1        -1.410840     -1.397387     -1.377214     -1.710116  \n",
       "2        -1.309691     -1.311862     -1.281074     -1.592582  \n",
       "3        -1.314860     -1.310705     -1.275450     -1.591451  \n",
       "4        -1.306474     -1.300084     -1.270758     -1.575720  \n",
       "...            ...           ...           ...           ...  \n",
       "5308     -0.774206     -0.822459     -0.711495     -1.229170  \n",
       "5309     -0.721160     -0.767271     -0.664786     -1.121473  \n",
       "5310     -0.727119     -0.774205     -0.693435     -1.150451  \n",
       "5311     -0.760749     -0.796523     -0.712018     -1.207490  \n",
       "5312      0.000000      0.000000      0.000000      0.000000  \n",
       "\n",
       "[5313 rows x 11 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/HomM1_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/Matrix_M1.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/Matrix_M1.csv.gz'\n",
    "replicates = 3\n",
    "wildtype(codon_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2e3fa190",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>rep_3</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>rep_3_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.588242e-06</td>\n",
       "      <td>-9.116109e-07</td>\n",
       "      <td>-1.015400e-06</td>\n",
       "      <td>-3.515253e-06</td>\n",
       "      <td>-0.558729</td>\n",
       "      <td>-0.165669</td>\n",
       "      <td>-0.130102</td>\n",
       "      <td>-0.583335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>1.172463e-02</td>\n",
       "      <td>1.176479e-03</td>\n",
       "      <td>-2.269012e-03</td>\n",
       "      <td>1.484895e-02</td>\n",
       "      <td>-0.547003</td>\n",
       "      <td>-0.164491</td>\n",
       "      <td>-0.132370</td>\n",
       "      <td>-0.568482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-3.077798e-06</td>\n",
       "      <td>-5.027946e-07</td>\n",
       "      <td>-2.219152e-07</td>\n",
       "      <td>-3.802507e-06</td>\n",
       "      <td>-0.558730</td>\n",
       "      <td>-0.165668</td>\n",
       "      <td>-0.130101</td>\n",
       "      <td>-0.583335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.890902e-09</td>\n",
       "      <td>-2.501504e-09</td>\n",
       "      <td>-1.469833e-09</td>\n",
       "      <td>-5.862239e-09</td>\n",
       "      <td>-0.558727</td>\n",
       "      <td>-0.165668</td>\n",
       "      <td>-0.130101</td>\n",
       "      <td>-0.583331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.590863e-09</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.590863e-09</td>\n",
       "      <td>-0.558727</td>\n",
       "      <td>-0.165668</td>\n",
       "      <td>-0.130101</td>\n",
       "      <td>-0.583331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10453</th>\n",
       "      <td>498</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.846334e-01</td>\n",
       "      <td>-1.332931e-01</td>\n",
       "      <td>-8.458195e-02</td>\n",
       "      <td>-3.361313e-01</td>\n",
       "      <td>-0.180615</td>\n",
       "      <td>-0.981435</td>\n",
       "      <td>-0.597915</td>\n",
       "      <td>-0.787924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10454</th>\n",
       "      <td>498</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-3.153976e-02</td>\n",
       "      <td>-4.990786e-02</td>\n",
       "      <td>-7.890274e-02</td>\n",
       "      <td>-1.561473e-01</td>\n",
       "      <td>-0.027521</td>\n",
       "      <td>-0.898050</td>\n",
       "      <td>-0.592235</td>\n",
       "      <td>-0.607940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10455</th>\n",
       "      <td>498</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-2.028449e-01</td>\n",
       "      <td>8.705869e-02</td>\n",
       "      <td>3.726587e-01</td>\n",
       "      <td>2.044085e-01</td>\n",
       "      <td>-0.198826</td>\n",
       "      <td>-0.761084</td>\n",
       "      <td>-0.140674</td>\n",
       "      <td>-0.247384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10456</th>\n",
       "      <td>498</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-2.480221e-01</td>\n",
       "      <td>4.274552e-02</td>\n",
       "      <td>3.133219e-02</td>\n",
       "      <td>-1.474319e-01</td>\n",
       "      <td>-0.244004</td>\n",
       "      <td>-0.805397</td>\n",
       "      <td>-0.482000</td>\n",
       "      <td>-0.599225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10457</th>\n",
       "      <td>498</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.575635e-01</td>\n",
       "      <td>-1.253935e-01</td>\n",
       "      <td>-8.649049e-02</td>\n",
       "      <td>-3.223271e-01</td>\n",
       "      <td>-0.153545</td>\n",
       "      <td>-0.973536</td>\n",
       "      <td>-0.599823</td>\n",
       "      <td>-0.774120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10458 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator         rep_1         rep_2  \\\n",
       "0         1          A         False -1.588242e-06 -9.116109e-07   \n",
       "1         1          R         False  1.172463e-02  1.176479e-03   \n",
       "2         1          N         False -3.077798e-06 -5.027946e-07   \n",
       "3         1          D         False -1.890902e-09 -2.501504e-09   \n",
       "4         1          C         False  0.000000e+00 -2.590863e-09   \n",
       "...     ...        ...           ...           ...           ...   \n",
       "10453   498          T         False -1.846334e-01 -1.332931e-01   \n",
       "10454   498          W         False -3.153976e-02 -4.990786e-02   \n",
       "10455   498          Y         False -2.028449e-01  8.705869e-02   \n",
       "10456   498          V         False -2.480221e-01  4.274552e-02   \n",
       "10457   498          *         False -1.575635e-01 -1.253935e-01   \n",
       "\n",
       "              rep_3         joint  rep_1_gauged  rep_2_gauged  rep_3_gauged  \\\n",
       "0     -1.015400e-06 -3.515253e-06     -0.558729     -0.165669     -0.130102   \n",
       "1     -2.269012e-03  1.484895e-02     -0.547003     -0.164491     -0.132370   \n",
       "2     -2.219152e-07 -3.802507e-06     -0.558730     -0.165668     -0.130101   \n",
       "3     -1.469833e-09 -5.862239e-09     -0.558727     -0.165668     -0.130101   \n",
       "4      0.000000e+00 -2.590863e-09     -0.558727     -0.165668     -0.130101   \n",
       "...             ...           ...           ...           ...           ...   \n",
       "10453 -8.458195e-02 -3.361313e-01     -0.180615     -0.981435     -0.597915   \n",
       "10454 -7.890274e-02 -1.561473e-01     -0.027521     -0.898050     -0.592235   \n",
       "10455  3.726587e-01  2.044085e-01     -0.198826     -0.761084     -0.140674   \n",
       "10456  3.133219e-02 -1.474319e-01     -0.244004     -0.805397     -0.482000   \n",
       "10457 -8.649049e-02 -3.223271e-01     -0.153545     -0.973536     -0.599823   \n",
       "\n",
       "       joint_gauged  \n",
       "0         -0.583335  \n",
       "1         -0.568482  \n",
       "2         -0.583335  \n",
       "3         -0.583331  \n",
       "4         -0.583331  \n",
       "...             ...  \n",
       "10453     -0.787924  \n",
       "10454     -0.607940  \n",
       "10455     -0.247384  \n",
       "10456     -0.599225  \n",
       "10457     -0.774120  \n",
       "\n",
       "[10458 rows x 11 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/PR8_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/PR8.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/PR8.csv.gz'\n",
    "replicates = 3\n",
    "wildtype(codon_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9b7b0e08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>1.260099e-04</td>\n",
       "      <td>-1.022034e-07</td>\n",
       "      <td>1.277259e-04</td>\n",
       "      <td>-0.037363</td>\n",
       "      <td>-0.031884</td>\n",
       "      <td>-0.066734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>4.552973e-04</td>\n",
       "      <td>7.052332e-04</td>\n",
       "      <td>1.197320e-03</td>\n",
       "      <td>-0.037034</td>\n",
       "      <td>-0.031179</td>\n",
       "      <td>-0.065664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-6.355282e-08</td>\n",
       "      <td>-4.109946e-08</td>\n",
       "      <td>-1.046523e-07</td>\n",
       "      <td>-0.037490</td>\n",
       "      <td>-0.031884</td>\n",
       "      <td>-0.066861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-5.827963e-10</td>\n",
       "      <td>-5.827963e-10</td>\n",
       "      <td>-0.037489</td>\n",
       "      <td>-0.031884</td>\n",
       "      <td>-0.066861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>-3.727336e-10</td>\n",
       "      <td>-3.885309e-10</td>\n",
       "      <td>-7.612645e-10</td>\n",
       "      <td>-0.037489</td>\n",
       "      <td>-0.031884</td>\n",
       "      <td>-0.066861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10453</th>\n",
       "      <td>498</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>4.667052e-02</td>\n",
       "      <td>4.290091e-02</td>\n",
       "      <td>9.936530e-02</td>\n",
       "      <td>-0.442375</td>\n",
       "      <td>-0.152895</td>\n",
       "      <td>-0.334225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10454</th>\n",
       "      <td>498</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-5.076745e-02</td>\n",
       "      <td>-6.087270e-02</td>\n",
       "      <td>-1.009676e-01</td>\n",
       "      <td>-0.539813</td>\n",
       "      <td>-0.256668</td>\n",
       "      <td>-0.534558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10455</th>\n",
       "      <td>498</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-4.273999e-02</td>\n",
       "      <td>-4.471882e-02</td>\n",
       "      <td>-7.991174e-02</td>\n",
       "      <td>-0.531785</td>\n",
       "      <td>-0.240514</td>\n",
       "      <td>-0.513502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10456</th>\n",
       "      <td>498</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.273709e-01</td>\n",
       "      <td>-1.342239e-01</td>\n",
       "      <td>-2.139638e-01</td>\n",
       "      <td>-0.616416</td>\n",
       "      <td>-0.330019</td>\n",
       "      <td>-0.647554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10457</th>\n",
       "      <td>498</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>3.831856e-02</td>\n",
       "      <td>-7.300603e-02</td>\n",
       "      <td>-2.379879e-02</td>\n",
       "      <td>-0.450727</td>\n",
       "      <td>-0.268802</td>\n",
       "      <td>-0.457389</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10458 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator         rep_1         rep_2  \\\n",
       "0         1          A         False  1.260099e-04 -1.022034e-07   \n",
       "1         1          R         False  4.552973e-04  7.052332e-04   \n",
       "2         1          N         False -6.355282e-08 -4.109946e-08   \n",
       "3         1          D         False  0.000000e+00 -5.827963e-10   \n",
       "4         1          C         False -3.727336e-10 -3.885309e-10   \n",
       "...     ...        ...           ...           ...           ...   \n",
       "10453   498          T         False  4.667052e-02  4.290091e-02   \n",
       "10454   498          W         False -5.076745e-02 -6.087270e-02   \n",
       "10455   498          Y         False -4.273999e-02 -4.471882e-02   \n",
       "10456   498          V         False -1.273709e-01 -1.342239e-01   \n",
       "10457   498          *         False  3.831856e-02 -7.300603e-02   \n",
       "\n",
       "              joint  rep_1_gauged  rep_2_gauged  joint_gauged  \n",
       "0      1.277259e-04     -0.037363     -0.031884     -0.066734  \n",
       "1      1.197320e-03     -0.037034     -0.031179     -0.065664  \n",
       "2     -1.046523e-07     -0.037490     -0.031884     -0.066861  \n",
       "3     -5.827963e-10     -0.037489     -0.031884     -0.066861  \n",
       "4     -7.612645e-10     -0.037489     -0.031884     -0.066861  \n",
       "...             ...           ...           ...           ...  \n",
       "10453  9.936530e-02     -0.442375     -0.152895     -0.334225  \n",
       "10454 -1.009676e-01     -0.539813     -0.256668     -0.534558  \n",
       "10455 -7.991174e-02     -0.531785     -0.240514     -0.513502  \n",
       "10456 -2.139638e-01     -0.616416     -0.330019     -0.647554  \n",
       "10457 -2.379879e-02     -0.450727     -0.268802     -0.457389  \n",
       "\n",
       "[10458 rows x 9 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/Aichi68C_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/Aichi68C.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/Aichi68C.csv.gz'\n",
    "replicates = 2\n",
    "wildtype(codon_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "809ef022",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>rep_3</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>rep_3_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>-5.774015e-10</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-1.802252e-08</td>\n",
       "      <td>-1.859993e-08</td>\n",
       "      <td>0.003398</td>\n",
       "      <td>0.013481</td>\n",
       "      <td>0.001531</td>\n",
       "      <td>0.016561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>-2.999165e-04</td>\n",
       "      <td>-2.999176e-04</td>\n",
       "      <td>-2.999207e-04</td>\n",
       "      <td>-8.997548e-04</td>\n",
       "      <td>0.003098</td>\n",
       "      <td>0.013181</td>\n",
       "      <td>0.001231</td>\n",
       "      <td>0.015661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.639706e-03</td>\n",
       "      <td>-2.761312e-03</td>\n",
       "      <td>-1.698832e-03</td>\n",
       "      <td>-6.117113e-03</td>\n",
       "      <td>0.001758</td>\n",
       "      <td>0.010720</td>\n",
       "      <td>-0.000168</td>\n",
       "      <td>0.010443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>-9.829320e-09</td>\n",
       "      <td>-1.988292e-08</td>\n",
       "      <td>-1.115037e-08</td>\n",
       "      <td>-4.086261e-08</td>\n",
       "      <td>0.003398</td>\n",
       "      <td>0.013481</td>\n",
       "      <td>0.001531</td>\n",
       "      <td>0.016560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-4.573932e-10</td>\n",
       "      <td>-4.573932e-10</td>\n",
       "      <td>0.003398</td>\n",
       "      <td>0.013481</td>\n",
       "      <td>0.001531</td>\n",
       "      <td>0.016561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11860</th>\n",
       "      <td>565</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>-2.269957e-03</td>\n",
       "      <td>-9.044772e-03</td>\n",
       "      <td>-8.256772e-03</td>\n",
       "      <td>-1.340381e-02</td>\n",
       "      <td>-0.251617</td>\n",
       "      <td>-0.285085</td>\n",
       "      <td>-0.282201</td>\n",
       "      <td>-0.585212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11861</th>\n",
       "      <td>565</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-9.687717e-03</td>\n",
       "      <td>-9.078947e-03</td>\n",
       "      <td>-8.186585e-03</td>\n",
       "      <td>-2.198810e-02</td>\n",
       "      <td>-0.259035</td>\n",
       "      <td>-0.285119</td>\n",
       "      <td>-0.282131</td>\n",
       "      <td>-0.593796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11862</th>\n",
       "      <td>565</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-8.586323e-03</td>\n",
       "      <td>-1.010091e-02</td>\n",
       "      <td>-8.385556e-03</td>\n",
       "      <td>-2.209933e-02</td>\n",
       "      <td>-0.257933</td>\n",
       "      <td>-0.286141</td>\n",
       "      <td>-0.282330</td>\n",
       "      <td>-0.593907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11863</th>\n",
       "      <td>565</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-4.195443e-03</td>\n",
       "      <td>-2.360678e-02</td>\n",
       "      <td>-1.729799e-02</td>\n",
       "      <td>-2.527836e-02</td>\n",
       "      <td>-0.253542</td>\n",
       "      <td>-0.299647</td>\n",
       "      <td>-0.291243</td>\n",
       "      <td>-0.597087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11864</th>\n",
       "      <td>565</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.298550e-02</td>\n",
       "      <td>-2.321162e-02</td>\n",
       "      <td>-1.241787e-02</td>\n",
       "      <td>-3.912170e-02</td>\n",
       "      <td>-0.262332</td>\n",
       "      <td>-0.299252</td>\n",
       "      <td>-0.286362</td>\n",
       "      <td>-0.610930</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11865 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       site amino_acid  WT_indicator         rep_1         rep_2  \\\n",
       "0         1          A         False -5.774015e-10  0.000000e+00   \n",
       "1         1          R         False -2.999165e-04 -2.999176e-04   \n",
       "2         1          N         False -1.639706e-03 -2.761312e-03   \n",
       "3         1          D         False -9.829320e-09 -1.988292e-08   \n",
       "4         1          C         False  0.000000e+00  0.000000e+00   \n",
       "...     ...        ...           ...           ...           ...   \n",
       "11860   565          T         False -2.269957e-03 -9.044772e-03   \n",
       "11861   565          W         False -9.687717e-03 -9.078947e-03   \n",
       "11862   565          Y         False -8.586323e-03 -1.010091e-02   \n",
       "11863   565          V         False -4.195443e-03 -2.360678e-02   \n",
       "11864   565          *         False -1.298550e-02 -2.321162e-02   \n",
       "\n",
       "              rep_3         joint  rep_1_gauged  rep_2_gauged  rep_3_gauged  \\\n",
       "0     -1.802252e-08 -1.859993e-08      0.003398      0.013481      0.001531   \n",
       "1     -2.999207e-04 -8.997548e-04      0.003098      0.013181      0.001231   \n",
       "2     -1.698832e-03 -6.117113e-03      0.001758      0.010720     -0.000168   \n",
       "3     -1.115037e-08 -4.086261e-08      0.003398      0.013481      0.001531   \n",
       "4     -4.573932e-10 -4.573932e-10      0.003398      0.013481      0.001531   \n",
       "...             ...           ...           ...           ...           ...   \n",
       "11860 -8.256772e-03 -1.340381e-02     -0.251617     -0.285085     -0.282201   \n",
       "11861 -8.186585e-03 -2.198810e-02     -0.259035     -0.285119     -0.282131   \n",
       "11862 -8.385556e-03 -2.209933e-02     -0.257933     -0.286141     -0.282330   \n",
       "11863 -1.729799e-02 -2.527836e-02     -0.253542     -0.299647     -0.291243   \n",
       "11864 -1.241787e-02 -3.912170e-02     -0.262332     -0.299252     -0.286362   \n",
       "\n",
       "       joint_gauged  \n",
       "0          0.016561  \n",
       "1          0.015661  \n",
       "2          0.010443  \n",
       "3          0.016560  \n",
       "4          0.016561  \n",
       "...             ...  \n",
       "11860     -0.585212  \n",
       "11861     -0.593796  \n",
       "11862     -0.593907  \n",
       "11863     -0.597087  \n",
       "11864     -0.610930  \n",
       "\n",
       "[11865 rows x 11 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_file = './data/raw_data/WSN_DNA_codoncounts.csv'\n",
    "selection_file = './output/selection_coefficients/WSN.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/WSN.csv.gz'\n",
    "replicates = 3\n",
    "wildtype(codon_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b943543",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "af6dad78",
   "metadata": {},
   "outputs": [],
   "source": [
    "def wildtype_full_length(sequence_file, selection_file, output_file, replicates):\n",
    "    with open(sequence_file, 'r') as file:\n",
    "        nucleotide = file.read().rstrip()\n",
    "        codon_list = [nucleotide[i:i+3] for i in range(0, len(nucleotide), 3)]\n",
    "        WT_aa = [codon2aa(i) for i in codon_list]\n",
    "\n",
    "    df_selection = pd.read_csv(selection_file)\n",
    "    site_list = df_selection['site'].unique()\n",
    "    df_WT = pd.DataFrame(columns = ['site', 'WT_aa', 'WT_indicator'])\n",
    "    df_WT['site'] = site_list\n",
    "    df_WT['WT_aa'] = WT_aa\n",
    "    df_WT['WT_indicator'] = True\n",
    "    # df_WT = df[['site', 'WT_aa','WT_indicator']]\n",
    "    df_WT = df_WT.rename(columns={'WT_aa': 'amino_acid'})\n",
    "\n",
    "    if 'gauged' not in ' '.join(df_selection.columns.tolist()):\n",
    "        df_merge = pd.merge(df_WT, df_selection, on=['site','amino_acid'], how='right')\n",
    "        df_merge['WT_indicator'] = df_merge['WT_indicator'].fillna(False)\n",
    "        column_name = df_merge.columns[-(replicates+1):].tolist()\n",
    "        gauged_name = [i+'_gauged' for i in column_name]\n",
    "        gauged_list = []\n",
    "        for i in gauged_name:\n",
    "            df_merge[i] = 0\n",
    "            gauged_list.append([])\n",
    "\n",
    "        for site in df_merge['site'].unique():\n",
    "            for i in range(len(column_name)):\n",
    "                gauge_value = df_merge[(df_merge['site']==site)&(df_merge['WT_indicator'])][column_name[i]].values[0]\n",
    "                original_list = df_merge[df_merge['site']==site][column_name[i]].tolist()\n",
    "    #                 print(gauge_value)\n",
    "                gauged_list[i] +=  [i - gauge_value for i in original_list]\n",
    "\n",
    "        for i in range(len(gauged_name)):\n",
    "            df_merge[gauged_name[i]] = gauged_list[i]\n",
    "\n",
    "        df_merge.to_csv(output_file, compression = 'gzip', index = False)\n",
    "        print('gauge done')\n",
    "        return df_merge\n",
    "    else:\n",
    "        print('gauge existed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5e8cc8f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>0.06310</td>\n",
       "      <td>0.086900</td>\n",
       "      <td>0.06450</td>\n",
       "      <td>0.24310</td>\n",
       "      <td>0.270900</td>\n",
       "      <td>0.25950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>0.05210</td>\n",
       "      <td>0.051000</td>\n",
       "      <td>0.08840</td>\n",
       "      <td>0.23210</td>\n",
       "      <td>0.235000</td>\n",
       "      <td>0.28340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.05600</td>\n",
       "      <td>-0.092200</td>\n",
       "      <td>-0.08870</td>\n",
       "      <td>0.12400</td>\n",
       "      <td>0.091800</td>\n",
       "      <td>0.10630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>D</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.18000</td>\n",
       "      <td>-0.184000</td>\n",
       "      <td>-0.19500</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00273</td>\n",
       "      <td>0.003270</td>\n",
       "      <td>0.00414</td>\n",
       "      <td>0.18273</td>\n",
       "      <td>0.187270</td>\n",
       "      <td>0.19914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>709</th>\n",
       "      <td>35</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00378</td>\n",
       "      <td>0.002820</td>\n",
       "      <td>0.00670</td>\n",
       "      <td>-0.38722</td>\n",
       "      <td>-0.391180</td>\n",
       "      <td>-0.39330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>710</th>\n",
       "      <td>35</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.39100</td>\n",
       "      <td>-0.394000</td>\n",
       "      <td>-0.40000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>711</th>\n",
       "      <td>35</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.01350</td>\n",
       "      <td>-0.011200</td>\n",
       "      <td>-0.02300</td>\n",
       "      <td>-0.40450</td>\n",
       "      <td>-0.405200</td>\n",
       "      <td>-0.42300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>712</th>\n",
       "      <td>35</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00244</td>\n",
       "      <td>0.000896</td>\n",
       "      <td>0.00347</td>\n",
       "      <td>-0.38856</td>\n",
       "      <td>-0.393104</td>\n",
       "      <td>-0.39653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713</th>\n",
       "      <td>35</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.00209</td>\n",
       "      <td>-0.001510</td>\n",
       "      <td>-0.00345</td>\n",
       "      <td>-0.39309</td>\n",
       "      <td>-0.395510</td>\n",
       "      <td>-0.40345</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>714 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     site amino_acid  WT_indicator    rep_1     rep_2    joint  rep_1_gauged  \\\n",
       "0       2          A         False  0.06310  0.086900  0.06450       0.24310   \n",
       "1       2          R         False  0.05210  0.051000  0.08840       0.23210   \n",
       "2       2          N         False -0.05600 -0.092200 -0.08870       0.12400   \n",
       "3       2          D          True -0.18000 -0.184000 -0.19500       0.00000   \n",
       "4       2          C         False  0.00273  0.003270  0.00414       0.18273   \n",
       "..    ...        ...           ...      ...       ...      ...           ...   \n",
       "709    35          T         False  0.00378  0.002820  0.00670      -0.38722   \n",
       "710    35          W         False  0.00000  0.000000  0.00000      -0.39100   \n",
       "711    35          Y         False -0.01350 -0.011200 -0.02300      -0.40450   \n",
       "712    35          V         False  0.00244  0.000896  0.00347      -0.38856   \n",
       "713    35          *         False -0.00209 -0.001510 -0.00345      -0.39309   \n",
       "\n",
       "     rep_2_gauged  joint_gauged  \n",
       "0        0.270900       0.25950  \n",
       "1        0.235000       0.28340  \n",
       "2        0.091800       0.10630  \n",
       "3        0.000000       0.00000  \n",
       "4        0.187270       0.19914  \n",
       "..            ...           ...  \n",
       "709     -0.391180      -0.39330  \n",
       "710     -0.394000      -0.40000  \n",
       "711     -0.405200      -0.42300  \n",
       "712     -0.393104      -0.39653  \n",
       "713     -0.395510      -0.40345  \n",
       "\n",
       "[714 rows x 9 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_file = './data/raw_data/YAP1_Reference_sequence.txt'\n",
    "selection_file = './output/selection_coefficients/YAP1.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/YAP1.csv.gz'\n",
    "replicates = 2\n",
    "\n",
    "wildtype_full_length(sequence_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "2b3bfad9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.1320</td>\n",
       "      <td>-0.09740</td>\n",
       "      <td>-0.1270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0274</td>\n",
       "      <td>-0.00606</td>\n",
       "      <td>0.0234</td>\n",
       "      <td>-0.1046</td>\n",
       "      <td>-0.10346</td>\n",
       "      <td>-0.1036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.1320</td>\n",
       "      <td>-0.09740</td>\n",
       "      <td>-0.1270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.1320</td>\n",
       "      <td>-0.09740</td>\n",
       "      <td>-0.1270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.1320</td>\n",
       "      <td>-0.09740</td>\n",
       "      <td>-0.1270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2158</th>\n",
       "      <td>104</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.0225</td>\n",
       "      <td>-0.01100</td>\n",
       "      <td>-0.0168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2159</th>\n",
       "      <td>104</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.0832</td>\n",
       "      <td>-0.06210</td>\n",
       "      <td>-0.0793</td>\n",
       "      <td>-0.1057</td>\n",
       "      <td>-0.07310</td>\n",
       "      <td>-0.0961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2160</th>\n",
       "      <td>104</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.0225</td>\n",
       "      <td>-0.01100</td>\n",
       "      <td>-0.0168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2161</th>\n",
       "      <td>104</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.0112</td>\n",
       "      <td>-0.01690</td>\n",
       "      <td>-0.0274</td>\n",
       "      <td>-0.0337</td>\n",
       "      <td>-0.02790</td>\n",
       "      <td>-0.0442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2162</th>\n",
       "      <td>104</td>\n",
       "      <td>*</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0225</td>\n",
       "      <td>0.01100</td>\n",
       "      <td>0.0168</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2163 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      site amino_acid  WT_indicator   rep_1    rep_2   joint  rep_1_gauged  \\\n",
       "0        2          A         False  0.0000  0.00000  0.0000       -0.1320   \n",
       "1        2          R         False  0.0274 -0.00606  0.0234       -0.1046   \n",
       "2        2          N         False  0.0000  0.00000  0.0000       -0.1320   \n",
       "3        2          D         False  0.0000  0.00000  0.0000       -0.1320   \n",
       "4        2          C         False  0.0000  0.00000  0.0000       -0.1320   \n",
       "...    ...        ...           ...     ...      ...     ...           ...   \n",
       "2158   104          T         False  0.0000  0.00000  0.0000       -0.0225   \n",
       "2159   104          W         False -0.0832 -0.06210 -0.0793       -0.1057   \n",
       "2160   104          Y         False  0.0000  0.00000  0.0000       -0.0225   \n",
       "2161   104          V         False -0.0112 -0.01690 -0.0274       -0.0337   \n",
       "2162   104          *          True  0.0225  0.01100  0.0168        0.0000   \n",
       "\n",
       "      rep_2_gauged  joint_gauged  \n",
       "0         -0.09740       -0.1270  \n",
       "1         -0.10346       -0.1036  \n",
       "2         -0.09740       -0.1270  \n",
       "3         -0.09740       -0.1270  \n",
       "4         -0.09740       -0.1270  \n",
       "...            ...           ...  \n",
       "2158      -0.01100       -0.0168  \n",
       "2159      -0.07310       -0.0961  \n",
       "2160      -0.01100       -0.0168  \n",
       "2161      -0.02790       -0.0442  \n",
       "2162       0.00000        0.0000  \n",
       "\n",
       "[2163 rows x 9 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_file = './data/raw_data/Ube4b_Reference_sequence.txt'\n",
    "selection_file = './output/selection_coefficients/Ube4b.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/Ube4b.csv.gz'\n",
    "replicates = 2\n",
    "\n",
    "wildtype_full_length(sequence_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "59a5c72f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>rep_3</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>rep_3_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>-0.000008</td>\n",
       "      <td>0.000084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.000050</td>\n",
       "      <td>-0.000055</td>\n",
       "      <td>-0.000043</td>\n",
       "      <td>-0.000153</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>-0.000028</td>\n",
       "      <td>-0.000050</td>\n",
       "      <td>-0.000069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>D</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.000073</td>\n",
       "      <td>-0.000028</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>-0.000084</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>-0.000008</td>\n",
       "      <td>0.000084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6358</th>\n",
       "      <td>304</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000299</td>\n",
       "      <td>-0.000337</td>\n",
       "      <td>-0.000288</td>\n",
       "      <td>-0.000900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6359</th>\n",
       "      <td>304</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000299</td>\n",
       "      <td>-0.000337</td>\n",
       "      <td>-0.000288</td>\n",
       "      <td>-0.000900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6360</th>\n",
       "      <td>304</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-0.000296</td>\n",
       "      <td>-0.000336</td>\n",
       "      <td>-0.000287</td>\n",
       "      <td>-0.000894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6361</th>\n",
       "      <td>304</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.000024</td>\n",
       "      <td>-0.000016</td>\n",
       "      <td>-0.000009</td>\n",
       "      <td>-0.000045</td>\n",
       "      <td>-0.000323</td>\n",
       "      <td>-0.000353</td>\n",
       "      <td>-0.000297</td>\n",
       "      <td>-0.000945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6362</th>\n",
       "      <td>304</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000299</td>\n",
       "      <td>-0.000337</td>\n",
       "      <td>-0.000288</td>\n",
       "      <td>-0.000900</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6363 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      site amino_acid  WT_indicator     rep_1     rep_2     rep_3     joint  \\\n",
       "0        2          A         False  0.000007  0.000009  0.000011  0.000027   \n",
       "1        2          R         False  0.000000  0.000000  0.000000  0.000000   \n",
       "2        2          N         False -0.000050 -0.000055 -0.000043 -0.000153   \n",
       "3        2          D          True -0.000073 -0.000028  0.000008 -0.000084   \n",
       "4        2          C         False  0.000000  0.000000  0.000000  0.000000   \n",
       "...    ...        ...           ...       ...       ...       ...       ...   \n",
       "6358   304          T         False  0.000000  0.000000  0.000000  0.000000   \n",
       "6359   304          W         False  0.000000  0.000000  0.000000  0.000000   \n",
       "6360   304          Y         False  0.000003  0.000001  0.000001  0.000006   \n",
       "6361   304          V         False -0.000024 -0.000016 -0.000009 -0.000045   \n",
       "6362   304          *         False  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "      rep_1_gauged  rep_2_gauged  rep_3_gauged  joint_gauged  \n",
       "0         0.000080      0.000036      0.000004      0.000111  \n",
       "1         0.000073      0.000028     -0.000008      0.000084  \n",
       "2         0.000024     -0.000028     -0.000050     -0.000069  \n",
       "3         0.000000      0.000000      0.000000      0.000000  \n",
       "4         0.000073      0.000028     -0.000008      0.000084  \n",
       "...            ...           ...           ...           ...  \n",
       "6358     -0.000299     -0.000337     -0.000288     -0.000900  \n",
       "6359     -0.000299     -0.000337     -0.000288     -0.000900  \n",
       "6360     -0.000296     -0.000336     -0.000287     -0.000894  \n",
       "6361     -0.000323     -0.000353     -0.000297     -0.000945  \n",
       "6362     -0.000299     -0.000337     -0.000288     -0.000900  \n",
       "\n",
       "[6363 rows x 11 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_file = './data/raw_data/BRCA1_Reference_sequence.txt'\n",
    "selection_file = './output/selection_coefficients/Y2H_1.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/Y2H_1.csv.gz'\n",
    "replicates = 3\n",
    "\n",
    "wildtype_full_length(sequence_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "65ff744c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>rep_3</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>rep_3_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.000003</td>\n",
       "      <td>-0.000006</td>\n",
       "      <td>-1.050000e-05</td>\n",
       "      <td>-0.000019</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.000274</td>\n",
       "      <td>0.000358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000134</td>\n",
       "      <td>0.000284</td>\n",
       "      <td>0.000377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.000096</td>\n",
       "      <td>-0.000064</td>\n",
       "      <td>-5.210000e-05</td>\n",
       "      <td>-0.000241</td>\n",
       "      <td>-0.000064</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000232</td>\n",
       "      <td>0.000136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>D</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.000032</td>\n",
       "      <td>-0.000134</td>\n",
       "      <td>-2.840000e-04</td>\n",
       "      <td>-0.000377</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.000134</td>\n",
       "      <td>0.000284</td>\n",
       "      <td>0.000377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6358</th>\n",
       "      <td>304</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000026</td>\n",
       "      <td>-0.000015</td>\n",
       "      <td>-0.000094</td>\n",
       "      <td>-0.000149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6359</th>\n",
       "      <td>304</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000026</td>\n",
       "      <td>-0.000015</td>\n",
       "      <td>-0.000094</td>\n",
       "      <td>-0.000149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6360</th>\n",
       "      <td>304</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>-1.130000e-06</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>-0.000025</td>\n",
       "      <td>-0.000011</td>\n",
       "      <td>-0.000095</td>\n",
       "      <td>-0.000144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6361</th>\n",
       "      <td>304</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.000006</td>\n",
       "      <td>-0.000005</td>\n",
       "      <td>-4.140000e-07</td>\n",
       "      <td>-0.000010</td>\n",
       "      <td>-0.000032</td>\n",
       "      <td>-0.000021</td>\n",
       "      <td>-0.000094</td>\n",
       "      <td>-0.000159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6362</th>\n",
       "      <td>304</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000026</td>\n",
       "      <td>-0.000015</td>\n",
       "      <td>-0.000094</td>\n",
       "      <td>-0.000149</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6363 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      site amino_acid  WT_indicator     rep_1     rep_2         rep_3  \\\n",
       "0        2          A         False -0.000003 -0.000006 -1.050000e-05   \n",
       "1        2          R         False  0.000000  0.000000  0.000000e+00   \n",
       "2        2          N         False -0.000096 -0.000064 -5.210000e-05   \n",
       "3        2          D          True -0.000032 -0.000134 -2.840000e-04   \n",
       "4        2          C         False  0.000000  0.000000  0.000000e+00   \n",
       "...    ...        ...           ...       ...       ...           ...   \n",
       "6358   304          T         False  0.000000  0.000000  0.000000e+00   \n",
       "6359   304          W         False  0.000000  0.000000  0.000000e+00   \n",
       "6360   304          Y         False  0.000001  0.000004 -1.130000e-06   \n",
       "6361   304          V         False -0.000006 -0.000005 -4.140000e-07   \n",
       "6362   304          *         False  0.000000  0.000000  0.000000e+00   \n",
       "\n",
       "         joint  rep_1_gauged  rep_2_gauged  rep_3_gauged  joint_gauged  \n",
       "0    -0.000019      0.000029      0.000128      0.000274      0.000358  \n",
       "1     0.000000      0.000032      0.000134      0.000284      0.000377  \n",
       "2    -0.000241     -0.000064      0.000070      0.000232      0.000136  \n",
       "3    -0.000377      0.000000      0.000000      0.000000      0.000000  \n",
       "4     0.000000      0.000032      0.000134      0.000284      0.000377  \n",
       "...        ...           ...           ...           ...           ...  \n",
       "6358  0.000000     -0.000026     -0.000015     -0.000094     -0.000149  \n",
       "6359  0.000000     -0.000026     -0.000015     -0.000094     -0.000149  \n",
       "6360  0.000005     -0.000025     -0.000011     -0.000095     -0.000144  \n",
       "6361 -0.000010     -0.000032     -0.000021     -0.000094     -0.000159  \n",
       "6362  0.000000     -0.000026     -0.000015     -0.000094     -0.000149  \n",
       "\n",
       "[6363 rows x 11 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_file = './data/raw_data/BRCA1_Reference_sequence.txt'\n",
    "selection_file = './output/selection_coefficients/Y2H_2.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/Y2H_2.csv.gz'\n",
    "replicates = 3\n",
    "\n",
    "wildtype_full_length(sequence_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "60449669",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>rep_3</th>\n",
       "      <th>rep_4</th>\n",
       "      <th>rep_5</th>\n",
       "      <th>rep_6</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>rep_3_gauged</th>\n",
       "      <th>rep_4_gauged</th>\n",
       "      <th>rep_5_gauged</th>\n",
       "      <th>rep_6_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.00198</td>\n",
       "      <td>-0.002430</td>\n",
       "      <td>-0.00137</td>\n",
       "      <td>-0.001020</td>\n",
       "      <td>-0.000787</td>\n",
       "      <td>-0.00107</td>\n",
       "      <td>-0.00848</td>\n",
       "      <td>-0.02048</td>\n",
       "      <td>-0.012930</td>\n",
       "      <td>-0.00324</td>\n",
       "      <td>-0.007280</td>\n",
       "      <td>-0.000214</td>\n",
       "      <td>-0.01957</td>\n",
       "      <td>-0.02538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.01850</td>\n",
       "      <td>-0.010500</td>\n",
       "      <td>-0.00187</td>\n",
       "      <td>-0.006260</td>\n",
       "      <td>0.000573</td>\n",
       "      <td>-0.01850</td>\n",
       "      <td>-0.01690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00403</td>\n",
       "      <td>-0.017200</td>\n",
       "      <td>-0.00392</td>\n",
       "      <td>0.011400</td>\n",
       "      <td>0.005000</td>\n",
       "      <td>0.00248</td>\n",
       "      <td>0.01860</td>\n",
       "      <td>-0.01447</td>\n",
       "      <td>-0.027700</td>\n",
       "      <td>-0.00579</td>\n",
       "      <td>0.005140</td>\n",
       "      <td>0.005573</td>\n",
       "      <td>-0.01602</td>\n",
       "      <td>0.00170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>D</td>\n",
       "      <td>True</td>\n",
       "      <td>0.01850</td>\n",
       "      <td>0.010500</td>\n",
       "      <td>0.00187</td>\n",
       "      <td>0.006260</td>\n",
       "      <td>-0.000573</td>\n",
       "      <td>0.01850</td>\n",
       "      <td>0.01690</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.01850</td>\n",
       "      <td>-0.010500</td>\n",
       "      <td>-0.00187</td>\n",
       "      <td>-0.006260</td>\n",
       "      <td>0.000573</td>\n",
       "      <td>-0.01850</td>\n",
       "      <td>-0.01690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6358</th>\n",
       "      <td>304</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.03130</td>\n",
       "      <td>0.006110</td>\n",
       "      <td>0.04200</td>\n",
       "      <td>0.028400</td>\n",
       "      <td>0.031000</td>\n",
       "      <td>0.03790</td>\n",
       "      <td>0.07240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6359</th>\n",
       "      <td>304</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.03130</td>\n",
       "      <td>0.006110</td>\n",
       "      <td>0.04200</td>\n",
       "      <td>0.028400</td>\n",
       "      <td>0.031000</td>\n",
       "      <td>0.03790</td>\n",
       "      <td>0.07240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6360</th>\n",
       "      <td>304</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00525</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.00380</td>\n",
       "      <td>-0.000104</td>\n",
       "      <td>0.002160</td>\n",
       "      <td>0.00233</td>\n",
       "      <td>0.00963</td>\n",
       "      <td>0.03655</td>\n",
       "      <td>0.006827</td>\n",
       "      <td>0.04580</td>\n",
       "      <td>0.028296</td>\n",
       "      <td>0.033160</td>\n",
       "      <td>0.04023</td>\n",
       "      <td>0.08203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6361</th>\n",
       "      <td>304</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00266</td>\n",
       "      <td>-0.000511</td>\n",
       "      <td>0.00115</td>\n",
       "      <td>0.001040</td>\n",
       "      <td>0.000888</td>\n",
       "      <td>0.00668</td>\n",
       "      <td>0.00818</td>\n",
       "      <td>0.03396</td>\n",
       "      <td>0.005599</td>\n",
       "      <td>0.04315</td>\n",
       "      <td>0.029440</td>\n",
       "      <td>0.031888</td>\n",
       "      <td>0.04458</td>\n",
       "      <td>0.08058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6362</th>\n",
       "      <td>304</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.03130</td>\n",
       "      <td>0.006110</td>\n",
       "      <td>0.04200</td>\n",
       "      <td>0.028400</td>\n",
       "      <td>0.031000</td>\n",
       "      <td>0.03790</td>\n",
       "      <td>0.07240</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6363 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      site amino_acid  WT_indicator    rep_1     rep_2    rep_3     rep_4  \\\n",
       "0        2          A         False -0.00198 -0.002430 -0.00137 -0.001020   \n",
       "1        2          R         False  0.00000  0.000000  0.00000  0.000000   \n",
       "2        2          N         False  0.00403 -0.017200 -0.00392  0.011400   \n",
       "3        2          D          True  0.01850  0.010500  0.00187  0.006260   \n",
       "4        2          C         False  0.00000  0.000000  0.00000  0.000000   \n",
       "...    ...        ...           ...      ...       ...      ...       ...   \n",
       "6358   304          T         False  0.00000  0.000000  0.00000  0.000000   \n",
       "6359   304          W         False  0.00000  0.000000  0.00000  0.000000   \n",
       "6360   304          Y         False  0.00525  0.000717  0.00380 -0.000104   \n",
       "6361   304          V         False  0.00266 -0.000511  0.00115  0.001040   \n",
       "6362   304          *         False  0.00000  0.000000  0.00000  0.000000   \n",
       "\n",
       "         rep_5    rep_6    joint  rep_1_gauged  rep_2_gauged  rep_3_gauged  \\\n",
       "0    -0.000787 -0.00107 -0.00848      -0.02048     -0.012930      -0.00324   \n",
       "1     0.000000  0.00000  0.00000      -0.01850     -0.010500      -0.00187   \n",
       "2     0.005000  0.00248  0.01860      -0.01447     -0.027700      -0.00579   \n",
       "3    -0.000573  0.01850  0.01690       0.00000      0.000000       0.00000   \n",
       "4     0.000000  0.00000  0.00000      -0.01850     -0.010500      -0.00187   \n",
       "...        ...      ...      ...           ...           ...           ...   \n",
       "6358  0.000000  0.00000  0.00000       0.03130      0.006110       0.04200   \n",
       "6359  0.000000  0.00000  0.00000       0.03130      0.006110       0.04200   \n",
       "6360  0.002160  0.00233  0.00963       0.03655      0.006827       0.04580   \n",
       "6361  0.000888  0.00668  0.00818       0.03396      0.005599       0.04315   \n",
       "6362  0.000000  0.00000  0.00000       0.03130      0.006110       0.04200   \n",
       "\n",
       "      rep_4_gauged  rep_5_gauged  rep_6_gauged  joint_gauged  \n",
       "0        -0.007280     -0.000214      -0.01957      -0.02538  \n",
       "1        -0.006260      0.000573      -0.01850      -0.01690  \n",
       "2         0.005140      0.005573      -0.01602       0.00170  \n",
       "3         0.000000      0.000000       0.00000       0.00000  \n",
       "4        -0.006260      0.000573      -0.01850      -0.01690  \n",
       "...            ...           ...           ...           ...  \n",
       "6358      0.028400      0.031000       0.03790       0.07240  \n",
       "6359      0.028400      0.031000       0.03790       0.07240  \n",
       "6360      0.028296      0.033160       0.04023       0.08203  \n",
       "6361      0.029440      0.031888       0.04458       0.08058  \n",
       "6362      0.028400      0.031000       0.03790       0.07240  \n",
       "\n",
       "[6363 rows x 17 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_file = './data/raw_data/BRCA1_Reference_sequence.txt'\n",
    "selection_file = './output/selection_coefficients/E3.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/E3.csv.gz'\n",
    "replicates = 6\n",
    "\n",
    "wildtype_full_length(sequence_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "3bd0eed2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>rep_3</th>\n",
       "      <th>rep_4</th>\n",
       "      <th>rep_5</th>\n",
       "      <th>rep_6</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>rep_3_gauged</th>\n",
       "      <th>rep_4_gauged</th>\n",
       "      <th>rep_5_gauged</th>\n",
       "      <th>rep_6_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>0.001220</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000280</td>\n",
       "      <td>0.000299</td>\n",
       "      <td>0.000541</td>\n",
       "      <td>0.00268</td>\n",
       "      <td>0.010540</td>\n",
       "      <td>0.005808</td>\n",
       "      <td>0.006356</td>\n",
       "      <td>0.011580</td>\n",
       "      <td>0.011799</td>\n",
       "      <td>0.014441</td>\n",
       "      <td>0.05558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.009320</td>\n",
       "      <td>0.005600</td>\n",
       "      <td>0.006320</td>\n",
       "      <td>0.011300</td>\n",
       "      <td>0.011500</td>\n",
       "      <td>0.013900</td>\n",
       "      <td>0.05290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>0.002130</td>\n",
       "      <td>0.000941</td>\n",
       "      <td>0.001120</td>\n",
       "      <td>0.003250</td>\n",
       "      <td>0.003720</td>\n",
       "      <td>0.002880</td>\n",
       "      <td>0.01230</td>\n",
       "      <td>0.011450</td>\n",
       "      <td>0.006541</td>\n",
       "      <td>0.007440</td>\n",
       "      <td>0.014550</td>\n",
       "      <td>0.015220</td>\n",
       "      <td>0.016780</td>\n",
       "      <td>0.06520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.009320</td>\n",
       "      <td>0.005600</td>\n",
       "      <td>0.006320</td>\n",
       "      <td>0.011300</td>\n",
       "      <td>0.011500</td>\n",
       "      <td>0.013900</td>\n",
       "      <td>0.05290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.009320</td>\n",
       "      <td>0.005600</td>\n",
       "      <td>0.006320</td>\n",
       "      <td>0.011300</td>\n",
       "      <td>0.011500</td>\n",
       "      <td>0.013900</td>\n",
       "      <td>0.05290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>32</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.006000</td>\n",
       "      <td>0.003990</td>\n",
       "      <td>0.004250</td>\n",
       "      <td>0.004030</td>\n",
       "      <td>0.005670</td>\n",
       "      <td>0.004530</td>\n",
       "      <td>0.02760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>32</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.006000</td>\n",
       "      <td>0.003990</td>\n",
       "      <td>0.004250</td>\n",
       "      <td>0.004030</td>\n",
       "      <td>0.005670</td>\n",
       "      <td>0.004530</td>\n",
       "      <td>0.02760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>32</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000362</td>\n",
       "      <td>0.000302</td>\n",
       "      <td>0.000202</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>0.000376</td>\n",
       "      <td>0.000138</td>\n",
       "      <td>0.00160</td>\n",
       "      <td>0.006362</td>\n",
       "      <td>0.004292</td>\n",
       "      <td>0.004452</td>\n",
       "      <td>0.004192</td>\n",
       "      <td>0.006046</td>\n",
       "      <td>0.004668</td>\n",
       "      <td>0.02920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>32</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000798</td>\n",
       "      <td>0.000473</td>\n",
       "      <td>0.000312</td>\n",
       "      <td>0.000897</td>\n",
       "      <td>0.000617</td>\n",
       "      <td>0.000823</td>\n",
       "      <td>0.00406</td>\n",
       "      <td>0.006798</td>\n",
       "      <td>0.004463</td>\n",
       "      <td>0.004562</td>\n",
       "      <td>0.004927</td>\n",
       "      <td>0.006287</td>\n",
       "      <td>0.005353</td>\n",
       "      <td>0.03166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>650</th>\n",
       "      <td>32</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.006000</td>\n",
       "      <td>0.003990</td>\n",
       "      <td>0.004250</td>\n",
       "      <td>0.004030</td>\n",
       "      <td>0.005670</td>\n",
       "      <td>0.004530</td>\n",
       "      <td>0.02760</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>651 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     site amino_acid  WT_indicator     rep_1     rep_2     rep_3     rep_4  \\\n",
       "0       2          A         False  0.001220  0.000208  0.000036  0.000280   \n",
       "1       2          R         False  0.000000  0.000000  0.000000  0.000000   \n",
       "2       2          N         False  0.002130  0.000941  0.001120  0.003250   \n",
       "3       2          D         False  0.000000  0.000000  0.000000  0.000000   \n",
       "4       2          C         False  0.000000  0.000000  0.000000  0.000000   \n",
       "..    ...        ...           ...       ...       ...       ...       ...   \n",
       "646    32          T         False  0.000000  0.000000  0.000000  0.000000   \n",
       "647    32          W         False  0.000000  0.000000  0.000000  0.000000   \n",
       "648    32          Y         False  0.000362  0.000302  0.000202  0.000162   \n",
       "649    32          V         False  0.000798  0.000473  0.000312  0.000897   \n",
       "650    32          *         False  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "        rep_5     rep_6    joint  rep_1_gauged  rep_2_gauged  rep_3_gauged  \\\n",
       "0    0.000299  0.000541  0.00268      0.010540      0.005808      0.006356   \n",
       "1    0.000000  0.000000  0.00000      0.009320      0.005600      0.006320   \n",
       "2    0.003720  0.002880  0.01230      0.011450      0.006541      0.007440   \n",
       "3    0.000000  0.000000  0.00000      0.009320      0.005600      0.006320   \n",
       "4    0.000000  0.000000  0.00000      0.009320      0.005600      0.006320   \n",
       "..        ...       ...      ...           ...           ...           ...   \n",
       "646  0.000000  0.000000  0.00000      0.006000      0.003990      0.004250   \n",
       "647  0.000000  0.000000  0.00000      0.006000      0.003990      0.004250   \n",
       "648  0.000376  0.000138  0.00160      0.006362      0.004292      0.004452   \n",
       "649  0.000617  0.000823  0.00406      0.006798      0.004463      0.004562   \n",
       "650  0.000000  0.000000  0.00000      0.006000      0.003990      0.004250   \n",
       "\n",
       "     rep_4_gauged  rep_5_gauged  rep_6_gauged  joint_gauged  \n",
       "0        0.011580      0.011799      0.014441       0.05558  \n",
       "1        0.011300      0.011500      0.013900       0.05290  \n",
       "2        0.014550      0.015220      0.016780       0.06520  \n",
       "3        0.011300      0.011500      0.013900       0.05290  \n",
       "4        0.011300      0.011500      0.013900       0.05290  \n",
       "..            ...           ...           ...           ...  \n",
       "646      0.004030      0.005670      0.004530       0.02760  \n",
       "647      0.004030      0.005670      0.004530       0.02760  \n",
       "648      0.004192      0.006046      0.004668       0.02920  \n",
       "649      0.004927      0.006287      0.005353       0.03166  \n",
       "650      0.004030      0.005670      0.004530       0.02760  \n",
       "\n",
       "[651 rows x 17 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_file = './data/raw_data/TpoR_Reference_sequence.txt'\n",
    "selection_file = './output/selection_coefficients/TpoR.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/TpoR.csv.gz'\n",
    "replicates = 6\n",
    "\n",
    "wildtype_full_length(sequence_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e7112473",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>WT_indicator</th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>rep_3</th>\n",
       "      <th>rep_4</th>\n",
       "      <th>rep_5</th>\n",
       "      <th>rep_6</th>\n",
       "      <th>joint</th>\n",
       "      <th>rep_1_gauged</th>\n",
       "      <th>rep_2_gauged</th>\n",
       "      <th>rep_3_gauged</th>\n",
       "      <th>rep_4_gauged</th>\n",
       "      <th>rep_5_gauged</th>\n",
       "      <th>rep_6_gauged</th>\n",
       "      <th>joint_gauged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.000112</td>\n",
       "      <td>-0.000128</td>\n",
       "      <td>0.001430</td>\n",
       "      <td>0.000504</td>\n",
       "      <td>0.000246</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>0.002090</td>\n",
       "      <td>-0.000308</td>\n",
       "      <td>-0.002118</td>\n",
       "      <td>0.004820</td>\n",
       "      <td>0.001934</td>\n",
       "      <td>0.001616</td>\n",
       "      <td>0.001745</td>\n",
       "      <td>0.007280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>R</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000196</td>\n",
       "      <td>-0.001990</td>\n",
       "      <td>0.003390</td>\n",
       "      <td>0.001430</td>\n",
       "      <td>0.001370</td>\n",
       "      <td>0.001570</td>\n",
       "      <td>0.005190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000251</td>\n",
       "      <td>-0.000101</td>\n",
       "      <td>0.000591</td>\n",
       "      <td>-0.000506</td>\n",
       "      <td>0.000484</td>\n",
       "      <td>0.000531</td>\n",
       "      <td>0.001210</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>-0.002091</td>\n",
       "      <td>0.003981</td>\n",
       "      <td>0.000924</td>\n",
       "      <td>0.001854</td>\n",
       "      <td>0.002101</td>\n",
       "      <td>0.006400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>D</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000196</td>\n",
       "      <td>-0.001990</td>\n",
       "      <td>0.003390</td>\n",
       "      <td>0.001430</td>\n",
       "      <td>0.001370</td>\n",
       "      <td>0.001570</td>\n",
       "      <td>0.005190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000196</td>\n",
       "      <td>-0.001990</td>\n",
       "      <td>0.003390</td>\n",
       "      <td>0.001430</td>\n",
       "      <td>0.001370</td>\n",
       "      <td>0.001570</td>\n",
       "      <td>0.005190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>32</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.002120</td>\n",
       "      <td>-0.000989</td>\n",
       "      <td>0.003790</td>\n",
       "      <td>-0.000605</td>\n",
       "      <td>0.000499</td>\n",
       "      <td>0.002290</td>\n",
       "      <td>0.002690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>32</td>\n",
       "      <td>W</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.002120</td>\n",
       "      <td>-0.000989</td>\n",
       "      <td>0.003790</td>\n",
       "      <td>-0.000605</td>\n",
       "      <td>0.000499</td>\n",
       "      <td>0.002290</td>\n",
       "      <td>0.002690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>32</td>\n",
       "      <td>Y</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.000375</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>0.000246</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>0.000136</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000279</td>\n",
       "      <td>-0.002495</td>\n",
       "      <td>-0.000900</td>\n",
       "      <td>0.004036</td>\n",
       "      <td>-0.000498</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.002368</td>\n",
       "      <td>0.002969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>32</td>\n",
       "      <td>V</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>-0.000020</td>\n",
       "      <td>-0.000006</td>\n",
       "      <td>-0.000177</td>\n",
       "      <td>0.000198</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.000250</td>\n",
       "      <td>-0.002026</td>\n",
       "      <td>-0.001009</td>\n",
       "      <td>0.003784</td>\n",
       "      <td>-0.000782</td>\n",
       "      <td>0.000697</td>\n",
       "      <td>0.002450</td>\n",
       "      <td>0.002940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>650</th>\n",
       "      <td>32</td>\n",
       "      <td>*</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.002120</td>\n",
       "      <td>-0.000989</td>\n",
       "      <td>0.003790</td>\n",
       "      <td>-0.000605</td>\n",
       "      <td>0.000499</td>\n",
       "      <td>0.002290</td>\n",
       "      <td>0.002690</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>651 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     site amino_acid  WT_indicator     rep_1     rep_2     rep_3     rep_4  \\\n",
       "0       2          A         False -0.000112 -0.000128  0.001430  0.000504   \n",
       "1       2          R         False  0.000000  0.000000  0.000000  0.000000   \n",
       "2       2          N         False  0.000251 -0.000101  0.000591 -0.000506   \n",
       "3       2          D         False  0.000000  0.000000  0.000000  0.000000   \n",
       "4       2          C         False  0.000000  0.000000  0.000000  0.000000   \n",
       "..    ...        ...           ...       ...       ...       ...       ...   \n",
       "646    32          T         False  0.000000  0.000000  0.000000  0.000000   \n",
       "647    32          W         False  0.000000  0.000000  0.000000  0.000000   \n",
       "648    32          Y         False -0.000375  0.000089  0.000246  0.000107   \n",
       "649    32          V         False  0.000094 -0.000020 -0.000006 -0.000177   \n",
       "650    32          *         False  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "        rep_5     rep_6     joint  rep_1_gauged  rep_2_gauged  rep_3_gauged  \\\n",
       "0    0.000246  0.000175  0.002090     -0.000308     -0.002118      0.004820   \n",
       "1    0.000000  0.000000  0.000000     -0.000196     -0.001990      0.003390   \n",
       "2    0.000484  0.000531  0.001210      0.000055     -0.002091      0.003981   \n",
       "3    0.000000  0.000000  0.000000     -0.000196     -0.001990      0.003390   \n",
       "4    0.000000  0.000000  0.000000     -0.000196     -0.001990      0.003390   \n",
       "..        ...       ...       ...           ...           ...           ...   \n",
       "646  0.000000  0.000000  0.000000     -0.002120     -0.000989      0.003790   \n",
       "647  0.000000  0.000000  0.000000     -0.002120     -0.000989      0.003790   \n",
       "648  0.000136  0.000078  0.000279     -0.002495     -0.000900      0.004036   \n",
       "649  0.000198  0.000160  0.000250     -0.002026     -0.001009      0.003784   \n",
       "650  0.000000  0.000000  0.000000     -0.002120     -0.000989      0.003790   \n",
       "\n",
       "     rep_4_gauged  rep_5_gauged  rep_6_gauged  joint_gauged  \n",
       "0        0.001934      0.001616      0.001745      0.007280  \n",
       "1        0.001430      0.001370      0.001570      0.005190  \n",
       "2        0.000924      0.001854      0.002101      0.006400  \n",
       "3        0.001430      0.001370      0.001570      0.005190  \n",
       "4        0.001430      0.001370      0.001570      0.005190  \n",
       "..            ...           ...           ...           ...  \n",
       "646     -0.000605      0.000499      0.002290      0.002690  \n",
       "647     -0.000605      0.000499      0.002290      0.002690  \n",
       "648     -0.000498      0.000635      0.002368      0.002969  \n",
       "649     -0.000782      0.000697      0.002450      0.002940  \n",
       "650     -0.000605      0.000499      0.002290      0.002690  \n",
       "\n",
       "[651 rows x 17 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_file = './data/raw_data/TpoR_Reference_sequence.txt'\n",
    "selection_file = './output/selection_coefficients/TpoR_S505N.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/TpoR_S505N.csv.gz'\n",
    "replicates = 6\n",
    "\n",
    "wildtype_full_length(sequence_file, selection_file, output_file, replicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5810bf64",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a7777a26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gauge done\n"
     ]
    }
   ],
   "source": [
    "excel_file = './data/raw_data/DBR1.xlsx'\n",
    "selection_file = './output/selection_coefficients/DBR1.csv.gz'\n",
    "output_file = './output/gauged_selection_coefficients/DBR1.csv.gz'\n",
    "replicates = 2\n",
    "\n",
    "df_selection = pd.read_csv(selection_file)\n",
    "site_list = df_selection['site'].unique()\n",
    "df_WT = pd.read_excel(excel_file)[['Affected codon', 'Reference AA']]\n",
    "df_WT = df_WT.drop_duplicates( keep='last').dropna()\n",
    "df_WT['WT_indicator'] = True\n",
    "# df_WT = df[['site', 'WT_aa','WT_indicator']]\n",
    "df_WT = df_WT.rename(columns={'Affected codon': 'site', 'Reference AA': 'amino_acid'})\n",
    "\n",
    "if 'gauged' not in ' '.join(df_selection.columns.tolist()):\n",
    "    df_merge = pd.merge(df_WT, df_selection, on=['site','amino_acid'], how='right')\n",
    "    df_merge['WT_indicator'] = df_merge['WT_indicator'].fillna(False)\n",
    "    column_name = df_merge.columns[-(replicates+1):].tolist()\n",
    "    gauged_name = [i+'_gauged' for i in column_name]\n",
    "    gauged_list = []\n",
    "    for i in gauged_name:\n",
    "        df_merge[i] = 0\n",
    "        gauged_list.append([])\n",
    "\n",
    "    for site in df_merge['site'].unique():\n",
    "        for i in range(len(column_name)):\n",
    "            gauge_value = df_merge[(df_merge['site']==site)&(df_merge['WT_indicator'])][column_name[i]].values[0]\n",
    "            original_list = df_merge[df_merge['site']==site][column_name[i]].tolist()\n",
    "            gauged_list[i] +=  [i - gauge_value for i in original_list]\n",
    "\n",
    "    for i in range(len(gauged_name)):\n",
    "        df_merge[gauged_name[i]] = gauged_list[i]\n",
    "\n",
    "    df_merge.to_csv(output_file, compression = 'gzip', index = False)\n",
    "    print('gauge done')\n",
    "else:\n",
    "    print('gauge existed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c928cf60",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
